---
title: "Clustering snRNAseq - combined model"
author: "Yuvarani Masarapu & Marta Graziano"
date: "2024-03-13"
output: 
  html_document:
    self_contained: true
    number_sections: true
    highlight: tango
    df_print: paged
    code_folding: hide
    toc: yes
    toc_depth: 3
    toc_float:
      collapsed: false
      smooth_scroll: true
---

```{r}
cat(paste(Sys.Date()))
```

# Load Libraries

```{r eval=FALSE}
#remotes::install_github('chris-mcginnis-ucsf/DoubletFinder')
#rmarkdown::render(input = "/Users/yuvarani.masarapu/Documents/parkinsons_project/sn_parkinsons_mitopark_combined/clustering_parkinsons_combined.Rmd", output_file = "/Users/yuvarani.masarapu/Documents/parkinsons_project/sn_parkinsons_mitopark_combined/clustering_combined.html")
```

```{r warning=FALSE, message=FALSE, eval=FALSE}
# install glmGamPoi

#if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
#BiocManager::install("glmGamPoi")
# install sctransform from Github
#install.packages("sctransform")
quartz<-function(width,height){windows(width, height)}
install.packages("devtools")
library(devtools)
remotes::install_version("Seurat", "4.4.0", repos = c("https://satijalab.r-universe.dev", getOption("repos")))
library(Seurat)
remotes::install_version("SeuratObject", "4.1.4", repos = c("https://satijalab.r-universe.dev", getOption("repos")))
library(SeuratObject)
install.packages("hdf5r")
library(hdf5r)
library(sctransform)
install.packages("harmony")
library(harmony)
remotes::install_github("chris-mcginnis-ucsf/DoubletFinder")
library(DoubletFinder)
library(patchwork)
library(Matrix)
library(ggplot2)
library(dplyr)
install.packages("openxlsx")
library(openxlsx)
# set this option when analyzing large datasets
options(future.globals.maxSize = 3e+09)
```

Custom function for selecting estimated doublet rate based on cells recovered from library

```{r doublet_finder, eval=FALSE}
run_doublet_finder <- function(x, do_plot = T, header = "") {
    # Given the number of cells in a library, return the estimated doublet rate. 
    # Number taken from
    # https://kb.10xgenomics.com/hc/en-us/articles/360001378811-What-is-the-maximum-number-of-cells-that-can-be-profiled-
    
    doublet_rate <- function(n_cells) {
        if (n_cells < 750) {
            return(0.004)
        }
        if (n_cells < 1500) {
            return(0.008)
        }
        if (n_cells < 2500) {
            return(0.016)
        }
        if (n_cells < 3500) {
            return(0.024)
        }
        if (n_cells < 4500) {
            return(0.032)
        }
        if (n_cells < 5500) {
            return(0.04)
        }
        if (n_cells < 6500) {
            return(0.048)
        }
        if (n_cells < 7500) {
            return(0.056)
        }
        if (n_cells < 8500) {
            return(0.064)
        }
        if (n_cells < 9500) {
            return(0.072)
        }
        if (n_cells < 10500) {
            return(0.08)
        }

        # extrapolating 0.01 for each additional 1000 cells.
        if (n_cells < 11500) {
            return(0.09)
        }
        if (n_cells < 12500) {
            return(0.1)
        }
        if (n_cells < 13500) {
            return(0.11)
        }
        if (n_cells < 13500) {
            return(0.12)
        }
        if (n_cells < 14500) {
            return(0.13)
        }
        if (n_cells < 15500) {
            return(0.14)
        }
        if (n_cells < 16500) {
            return(0.16)
        } 
        else (return(0.17))
    }

    # Normalize for doublet finder
    DefaultAssay(x) <- "RNA"
    x <- NormalizeData(x)
    x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
    x <- ScaleData(x)
    x <- RunPCA(x)
    x <- RunUMAP(x, dims = 1:30)

    # pK Identification (no ground-truth)
    sweep_list <- paramSweep(x, PCs = 1:10)
    sweep_stats <- summarizeSweep(sweep_list)
    bcmvny <- find.pK(sweep_stats)
    use_pK <- as.numeric(levels(bcmvny$pK)[which(bcmvny$BCmetric == max(bcmvny$BCmetric))])

    # Estimate doublet rate and predict doublets
    n_cells <- nrow(x@meta.data)
    nExp_poi <- round(doublet_rate(n_cells) * n_cells)
    x <- doubletFinder(x, PCs = 1:10, pN = 0.25, pK = use_pK, nExp = nExp_poi)

    DF.name = colnames(x@meta.data)[grepl("DF.classification", colnames(x@meta.data))]

    if (do_plot) {
        print(DimPlot(x, group.by = DF.name) + ggtitle(header))
        print(VlnPlot(x, features = "nFeature_RNA", group.by = DF.name, pt.size = 0))
    }

    out_pred <- x@meta.data[, DF.name]
    names(out_pred) <- colnames(x)
    return(x)
    # return(out_pred)
}
```

# Read count matrix into R as seurat object

Count data from all mice samples of mitopark and 6OHDA models are combined. The sample labels are replaced to new from old labels as below.

```{r load_data, message=FALSE, eval=FALSE, warning=FALSE}
library(tidyverse)
library(Matrix)
library(cowplot)
theme_set(theme_cowplot())
setwd("C:/Users/magraz/OneDrive - Karolinska Institutet/Parkinson Project-common folder")
indir <- "C:/Users/magraz/OneDrive - Karolinska Institutet/Parkinson Project-common folder/Counts_cellranger/"
data_dir <- paste(indir, "/cellranger-count_snRNAseq_6OHDA", sep = "")
files <- dir(data_dir)

files

seurat_list <- lapply(files, function(file_n){
  print(file_n)
  
  matrix_dir <- paste(data_dir, "/", file_n, "/outs/filtered_feature_bc_matrix/", sep = "")
  
  barcode.path <- paste0(matrix_dir, "barcodes.tsv.gz")
  features.path <- paste0(matrix_dir, "features.tsv.gz")
  matrix.path <- paste0(matrix_dir, "matrix.mtx.gz")
  mat <- readMM(file = matrix.path)
  
  feature.names = read.delim(features.path,
                           header = FALSE,
                           stringsAsFactors = FALSE)
  barcode.names = read.delim(barcode.path,
                           header = FALSE,
                           stringsAsFactors = FALSE)
  colnames(mat) = barcode.names$V1
  rownames(mat) = feature.names$V2 #we need gene names in form of ID, so we take column 2 where gene ids are
  
  # create seurat object
  cur_seurat <- CreateSeuratObject(
    counts = mat,
    project = "6OHDA"
  )
  cur_seurat@meta.data$SampleID <- file_n
  
  cur_seurat

})

new_sample_names <- c("15a", "15b", "15c", "15d", "18a", "18d", "C4", "C5")
sample_type <- c("Mitopark_15", "Mitopark_15", "Mitopark_15", "Mitopark_15", "Mitopark_18", "Mitopark_18", "Control", "Control")
sample_group <- c("Treated", "Treated", "Treated", "Treated", "Treated", "Treated", "Control", "Control")
for(i in 1:length(seurat_list)){
  seurat_list[[i]]$Sample <- new_sample_names[i]
  seurat_list[[i]]$Sample_type <- sample_type[i]
  seurat_list[[i]]$Sample_group <- sample_group[i]
}

gc()
```

```{r merge_data, warning=FALSE, message=FALSE, eval=FALSE}
# merge into one big seurat object
seurat_rna <- merge(x=seurat_list[[1]], y=seurat_list[2:length(seurat_list)])
rm(seurat_list)
```

```{r cellcyclescoring, message=FALSE, verbose=FALSE, eval=FALSE}
# Classify each cell by cell-cycle phase (G2M/G1/S) and give cellcycle scoring
seurat_rna_mp<- readRDS("C:/Users/magraz/OneDrive - Karolinska Institutet/Parkinson Project-common folder/analysis_results/sn_data/clustering_mitopark/data/raw_combined_mt_11_15_18.rds")
seurat_rna_6ohda<- readRDS("C:/Users/magraz/OneDrive - Karolinska Institutet/Parkinson Project-common folder/analysis_results/sn_data/clustering_6OHDA/data/raw_combined.rds")

#CHECK SAMPLE NAMES
IDS1<-unique(seurat_rna_mp$Sample)

IDS2<-unique(seurat_rna_6ohda$Sample)
print(IDS1)
print(IDS2)

seurat_rna <- merge(seurat_rna_mp, y = seurat_rna_6ohda,  project = "sRNA seq")
IDS<-unique(seurat_rna$Sample)
rm(seurat_rna_mp)
rm(seurat_rna_6ohda)

setwd("Z:/dmclab/Marta/Combined_snRNAseq")

indir <- getwd()

# A list of cell cycle markers, from Tirosh et al, 2015, is loaded with Seurat.  We can
# segregate this list into markers of G2/M phase and markers of S phase
# Format cell cycle genes from Seurat, e.g. convert 'MCM5' to 'Mcm5'
s.genes <- cc.genes$s.genes %>%
              str_to_lower %>%
              str_to_title
g2m.genes <- cc.genes$g2m.genes %>%
              str_to_lower %>%
              str_to_title

all.genes <- rownames(seurat_rna)

g2m.genes <- intersect(g2m.genes, all.genes)
s.genes <- intersect(s.genes, all.genes)

seurat_rna <- CellCycleScoring(object = seurat_rna, assay = "RNA", g2m.features = g2m.genes, s.features = s.genes)

saveRDS(seurat_rna, file = "Z:/dmclab/Marta/Combined_snRNAseq/data/raw_combined_snRNAseq.rds")
```

Let's check the size of data (nuclei count) for each sample

```{r data_size_plots, warning=FALSE, message=FALSE, fig.height=4, fig.width=4}
# use table function to get the number of cells in each Sample as a dataframe
seurat_rna$Sample_group <- colnames(seurat_rna)

seurat_rna$Sample_group[seurat_rna$Sample %in% c("15a", "15b", "15c","15d","18a", "18d")] <- "Mitopark15_18"
seurat_rna$Sample_group[seurat_rna$Sample %in% c("MP1", "MP2", "MP3")] <- "Mitopark_11"
seurat_rna$Sample_group[seurat_rna$Sample %in% c("C1", "C2", "C3", "C4", "C5", "Control_1", "Control_2" ,"Control_3")] <- "Control"
seurat_rna$Sample_group[seurat_rna$Sample %in% c("6OHDA_1", "6OHDA_2", "6OHDA_3")] <- "6OHDA"


# seurat_rna_tot <- readRDS(file = "C:/Users/magraz/OneDrive - Karolinska Institutet/Parkinson Project-common folder/analysis_results/sn_data/clustering_mitopark/data/raw_combined_mt_11_15_18.rds")
# 
# seurat_ctrl<- readRDS("C:/Users/magraz/OneDrive - Karolinska Institutet/snRNAsequencing data Striatum/control_dataset.rds")
# seurat_rna_tot$Sample[seurat_rna_tot$Sample %in% c("MP1", "MP2", "MP3")] <- c("11a","11b","11c")

df <- as.data.frame(table(rna.combined$Sample))
colnames(df) <- c('Sample', 'n_cells')

# bar plot of the number of cells in each sample
p <- ggplot(df, aes(y=n_cells, x=reorder(Sample, Sample), fill=Sample)) +
  geom_bar(stat='identity') +
  geom_text(aes(label = n_cells), vjust = 1.5, colour = "black") +
  scale_y_continuous(expand = c(0,0)) +
  NoLegend() + RotatedAxis() +
  ylab(expression(italic(N)[cells])) + xlab('Sample Name') +
  ggtitle(paste('Total cells:', sum(df$n_cells))) +
  theme_classic()+
  theme(legend.position = "none")
  plot_filename <- paste0("Z:/dmclab/Marta/Combined_snRNAseq/stat_plots/barplot_n_afterfilt", ".pdf")
  pdf(plot_filename, width = 9, height = 4)  # Adjust width and height as desired
  print(p)
  dev.off()
#png('figures/basic_cells_per_sample.png', width=9, height=4, res=200, units='in')

p
```

```{r data_size_plots, warning=FALSE, message=FALSE, fig.height=4, fig.width=4}
df1 <- as.data.frame(table(seurat_rna$Sample_group))
colnames(df1) <- c('Sample_group', 'n_cells')

# bar plot of the number of cells in each sample
p1 <- ggplot(df1, aes(y=n_cells, x=reorder(Sample_group, Sample_group), fill=Sample_group)) +
  geom_bar(stat='identity') +
  geom_text(aes(label = n_cells), vjust = 1.5, colour = "black") +
  scale_y_continuous(expand = c(0,0)) +
  NoLegend() + RotatedAxis() +
  ylab(expression(italic(N)[cells])) + xlab('Sample Group') +
  ggtitle(paste('Total cells:', sum(df$n_cells))) +
  theme(
    panel.grid.minor=element_blank(),
    panel.grid.major.y=element_line(colour="lightgray", size=0.5),
  )

#png('figures/basic_cells_per_sample.png', width=9, height=4, res=200, units='in')

p1
```

```         
```

# Before filtering

```{r most_expressed_genes, warning=FALSE, message=FALSE, eval=FALSE}
## Top 20 most expressed genes (most abundant transcripts)

par(mar = c(4, 8, 2, 1))
C <- seurat_rna@assays$RNA@counts
C <- Matrix::t(Matrix::t(C)/Matrix::colSums(C)) * 100
saveRDS(C, file = "C.rds")
most_expressed <- order(apply(C, 1, median), decreasing = T)[20:1]
saveRDS(most_expressed, file = "most_expressed.rds")
```

```{r most_expressed_plot, warning=FALSE, message=FALSE, fig.height=5, fig.width=5, eval=FALSE}
# Plot most abundant transctipts
par(mar = c(4, 8, 2, 1))
C <- readRDS("C.rds")
most_expressed <- readRDS("most_expressed.rds")
b<-boxplot(as.matrix(t(C[most_expressed, ])), cex = 0.05, las = 1, xlab = "% total count per cell", col = (scales::hue_pal())(20)[20:1], horizontal = TRUE)

rm(p)
```

## Violin plots with raw values

```{r violins_raw_bef_filt, fig.height=4, fig.width=10}
# removes epsilon values and plots real values instead!!!!
# https://stackoverflow.com/questions/18600115/forcing-a-1e3-instead-of-1000-format-in-ggplot-r/18600721#18600721
options(scipen=200000)

p1 <- VlnPlot(object = seurat_rna, features = c("nFeature_RNA"), group.by = "Sample",
    pt.size = 0, log = F)

p2 <- VlnPlot(object = seurat_rna, features = c("nCount_RNA"), group.by = "Sample",
    pt.size = 0, log = F)

p3<-cowplot::plot_grid(plotlist = list(p1,p2), nrow = 1, ncol = 2)
  plot_filename <- paste0(plot_directory, "/violin_beforefilt", ".pdf")
  pdf(plot_filename, width = 18, height = 8)  # Adjust width and height as desired
  print(p3)
  dev.off()
```

## Violin plots with log scaled values

```{r violins_log_bef_filt, fig.height=4, fig.width=10}
p1 <- VlnPlot(object = seurat_rna, features = c("nFeature_RNA"), group.by = "Sample",
    pt.size = 0, log = T)

p2 <- VlnPlot(object = seurat_rna, features = c("nCount_RNA"), group.by = "Sample",
    pt.size = 0, log = T)

p3<-cowplot::plot_grid(plotlist = list(p1,p2), nrow = 1, ncol = 2)
  plot_filename <- paste0(plot_directory, "/violin_beforefilt_log", ".pdf")
  pdf(plot_filename, width = 18, height = 8)  # Adjust width and height as desired
  print(p3)
  dev.off()
rm(p1,p2)
```

# After filtering

-   Overexpressed Malat1 gene, \n
-   mitochondrial and ribosomal genes \n
-   All genes with 0 transcripts detected
-   nFeature [250, 10000]
-   nCount [300, 75000]
-   Genes per UMI per cell ratio above 0.82 (gives an idea of the complexity of each cell)

```{r filtering, warning=FALSE}
#https://bookdown.org/ytliu13207/SingleCellMultiOmicsDataAnalysis/seurat-qc-cell-level-filtering.html

#remove counts variable from previous step
rm(C)

#Remove MALAT1 and x related genes
seurat_rna <- seurat_rna[!grepl(c("Malat1","Xist", "Uty", "Tsix",'Ddx3y', "Ddx3x", "Tbl1x","Usp9x","Jpx","Ftx","Tsnax","Pcdh11x", 'Eif2s3y',"Eif2s3x","Tmsb4x","Tmsb4y", 'Kdm5d'), rownames(seurat_rna))]

# Remove mitochondrial genes, 13 genes filtered out
seurat_rna[["percent_mt"]] <- PercentageFeatureSet(seurat_rna, pattern = "^mt-")
seurat_rna <- seurat_rna[!grepl("^mt-", rownames(seurat_rna)), ]

# Remove ribosomal genes, 111 genes filtered out
seurat_rna <- seurat_rna[!grepl("^Rps|^Rpl", rownames(seurat_rna)), ]

# log10 of number of genes detected per UMI
seurat_rna$log10GenesPerUMI <- log10(seurat_rna$nFeature_RNA) / log10(seurat_rna$nCount_RNA)
```

```{r data.table_filteing, message=FALSE, warning=FALSE}

seurat_rna[["orig.ident"]] <- "snRNA seq"
df <- as.data.table(seurat_rna@meta.data)

sel <- c("orig.ident", "nCount_RNA", "nFeature_RNA", "percent_mt", "log10GenesPerUMI")
df <- df[, sel, with = FALSE]
#df[1:3, ]
```

```{r plots_after_filtering, fig.height=7, fig.width=7}


fontsize <- 10
linesize <- 0.35

gp.ls <- df[, 2:5] %>% imap( ~ {
  
   # define lable fun
  give.n <- function(x) {
    return(c(y = median(x) + max(x) / 10, label = round(median(x), 2)))
  }
  
  # assign colors
  col.ls <-
    setNames(
      c('lightpink2', 'lightblue2', 'lightgreen', 'coral1'),
      c("nCount_RNA", "nFeature_RNA", "percent_mt", "log10GenesPerUMI")
    )
  
  ggplot(data = df, aes(x = orig.ident, y = .x)) +
    geom_violin(trim = FALSE, fill = col.ls[.y]) +
    ggtitle(label = .y) + ylab(label = .y) +
    theme_bw() +
    theme(
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      strip.background = element_blank(),
      panel.border = element_blank()
    ) +
    theme(
      axis.text = element_text(size = fontsize),
      axis.line = element_line(colour = "black", size = linesize),
      axis.ticks = element_line(size = linesize),
      axis.title.x = element_blank(),
      axis.ticks.length = unit(.05, "cm"),
      plot.title = element_text(size = fontsize + 2, hjust = 0.5),
      legend.position = 'none'
    ) +
    stat_summary(fun = median, geom = "point", col = "black") +  # Add points to plot
    stat_summary(fun.data = give.n,
                 geom = "text",
                 col = "black")
})

 grid.arrange(gp.ls[[1]], gp.ls[[2]], gp.ls[[3]], gp.ls[[4]], ncol = 2)
```

```{r filtering_stats, message=FALSE, warning=FALSE, fig.height=4, fig.width=6}
# Add metadata back to Seurat object
metadata <- seurat_rna@meta.data

# Visualize the number UMIs/transcripts per cell
metadata %>% 
  	ggplot(aes(color=Sample, x=nCount_RNA, fill= Sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  labs(title="Visualize the number UMIs/transcripts per cell")+
  	geom_vline(xintercept = c(400, 75000))

# Visualize the number genes per cell
metadata %>% 
  	ggplot(aes(color=Sample, x=nFeature_RNA, fill= Sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  labs(title=" Visualize the number genes per cell")+
  	geom_vline(xintercept = c(300, 10000))

# Visualize the number genes per cell
metadata %>% 
  	ggplot(aes(color=Sample, x=log10GenesPerUMI, fill= Sample)) + 
  	geom_density(alpha = 0.2) + 
  	scale_x_log10() + 
  	theme_classic() +
  	ylab("Cell density") +
  	geom_vline(xintercept = c(0.82))
```

```{r subsetting, warning=FALSE}
# subset will do cell-level filtering, remove cells with detected genes less than 350
seurat_rna <- subset(seurat_rna, subset = nFeature_RNA > 300 & nFeature_RNA < 10000 & nCount_RNA > 400 &  nCount_RNA < 75000)

# complexity of each cell to be > 0.80, so genes detected proportional to UMI
seurat_rna <- subset(seurat_rna, subset = log10GenesPerUMI > 0.82)

# Always do gene level threshold filtering AFTER cell filtering. 
# There could be some rows or genes with '0' transcripts that can remain after the cell filtering and if done before, this may lead to some rows in the matrix with all '0' values across.
# We also want to remove genes with 0 transcripts. Using subset() will not work here
seurat_rna <- seurat_rna[rowSums(seurat_rna) > 0, ]
```

# Check data quality after filtering

### Y-axis with raw values

```{r violins_raw_after_filt, fig.height=4, fig.width=10}
p1 <- VlnPlot(object = seurat_rna, features = c("nFeature_RNA"), group.by = "Sample",
    pt.size = 0, log = F)

p2 <- VlnPlot(object = seurat_rna, features = c("nCount_RNA"), group.by = "Sample",
    pt.size = 0, log = F)

p3<-cowplot::plot_grid(plotlist = list(p1,p2), nrow = 1, ncol = 2)
  plot_filename <- paste0(plot_directory, "/violin_afterfilt", ".pdf")
  pdf(plot_filename, width = 18, height = 8)  # Adjust width and height as desired
  print(p3)
  dev.off()
```

### Y-axis with log scale

```{r violins_log_after_filt, fig.height=4, fig.width=10}
p1 <- VlnPlot(object = seurat_rna, features = c("nFeature_RNA"), group.by = "Sample",
    pt.size = 0, log = T)

p2 <- VlnPlot(object = seurat_rna, features = c("nCount_RNA"), group.by = "Sample",
    pt.size = 0, log = T)

p3<-cowplot::plot_grid(plotlist = list(p1,p2), nrow = 1, ncol = 2)
  plot_filename <- paste0(plot_directory, "/violin_afterfilt_log", ".pdf")
  pdf(plot_filename, width = 18, height = 8)  # Adjust width and height as desired
  print(p3)
  dev.off()
rm(p1,p2)
```

### Percentage mitchondria per nuclei

-   Removed all mt- genes but let's see how they were distributed across each cell/nuclei

```{r violins_mt, fig.height=5, fig.width=6}
v<-VlnPlot(object = seurat_rna, features = c("percent_mt"), group.by = "Sample",
    pt.size = 0, log = T)
```

```{r message=FALSE}
rm(seurat_rna_tot)
```

# Run doubletfinder

-   finds doublets, removes them and also normalizes the data

```{r split_obj, eval=FALSE}
seurat_rna_list <- SplitObject(seurat_rna, split.by = "Sample")

names(seurat_rna_list)


```

```{r find_doublets, warning=FALSE, message=FALSE, verbose = FALSE, eval=FALSE}

set.seed(1)

all_doublets <- NULL
all_doublet_scores <- NULL
pANN_max <- 1

for (x_name in names(seurat_rna_list)){
  
    x <- seurat_rna_list[[x_name]]

    VlnPlot(x, features = c("nFeature_RNA", "nCount_RNA", "percent_mt"), ncol = 3,
        log = T, pt.size = 0) + NoLegend()
    
    x <- run_doublet_finder(x, do_plot = T, header = x_name)
    doublet_cutoff <- 0.6
    pANN.name = colnames(x@meta.data)[grepl("pANN", colnames(x@meta.data))]
    DF.name = colnames(x@meta.data)[grepl("DF", colnames(x@meta.data))]
    new_doublet_scores <- x[[pANN.name]][, 1]
    names(new_doublet_scores) <- rownames(x[[pANN.name]])
    new_doublet_calls <- x[[DF.name]][, 1]
    names(new_doublet_calls) <- rownames(x[[DF.name]])
    
    #final cells which are labelled doublets
    new_doublets <- union(names(which(new_doublet_calls == "Doublet")), names(which(new_doublet_scores >=
        pANN_max)))
    x <- x[, setdiff(colnames(x), new_doublets)]
    all_doublets <- union(all_doublets, new_doublets)
    all_doublet_scores <- c(all_doublet_scores, new_doublet_scores)
    
    #filtering step again to remove rows with '0' transcripts
    x <- x[rowSums(x) > 0, ]
    
    #Cellcyle scoring of each cell
    #x <- CellCycleScoring(object = x, assay = "SCT", g2m.features = g2m.genes, s.features = s.genes)
    
    # adding vst.flavour = "v2" makes values 0 across rows for some genes.
    # x <- SCTransform(x, verbose = T, variable.features.n = 6000, vars.to.regress = c("S.Score", "G2M.Score"))
    x <- SCTransform(x, verbose = T, variable.features.n = 6000) #25 perc of variable genes

    seurat_rna_list[[x_name]] <- x
}

rm(x)

for(i in 1:length(seurat_rna_list)){
  seurat_rna_list[[i]] <- AddMetaData(
  object = seurat_rna_list[[i]] ,
  metadata = all_doublet_scores[colnames(seurat_rna_list[[i]])],
  col.name = 'pANN'
)
}

saveRDS(seurat_rna_list, file = "seurat_rna_list_doublets.rds")
gc()
saveRDS(all_doublet_scores, file = "all_doublet_scores.rds")

```

Downstream processing

```{r eval=FALSE}
# select features that are repeatedly variable across datasets for integration
# Taking the top 25% of the total

# features <- readRDS("C:/Users/magraz/OneDrive - Karolinska Institutet/Parkinson Project-common folder/analysis_results/sn_data/combined_clustering/data/features.rds")
# rna.anchors <- readRDS("C:/Users/magraz/OneDrive - Karolinska Institutet/Parkinson Project-common folder/analysis_results/sn_data/combined_clustering/rna.anchors.rds")
#seurat_rna_list<-readRDS("C:/Users/dmc-analysis/Desktop/Marta/seurat_rna_list_doublets.rds")


#CHECK DIFFERENT nfEATURES AND CHECK WHEN IT STOPS AND THEN TAKE THE TOP 50% VARIABLE GENES OR WHATEVER PERCENTAGE YOU WANT:
#Max variable features foud: 9242
features <- SelectIntegrationFeatures(object.list = seurat_rna_list, nfeatures = 6000)

saveRDS(features, file = "Z:/dmclab/Marta/Combined_snRNAseq/data/features_used.rds")
seurat_rna_list <- PrepSCTIntegration(object.list = seurat_rna_list, anchor.features = features)


 
max_ncol <- 0
max_seurat <- NULL
for (i in seq_along(seurat_rna_list)) {
    seurat_rna <- seurat_rna_list[[i]]  # Get the Seurat object at index i
    
    # Get the number of columns for the current Seurat object
    current_ncol <- ncol(seurat_rna)
    
    # Check if the number of columns for the current Seurat object is greater than the maximum number of columns encountered so far
    if (current_ncol > max_ncol) {
        # Update the maximum number of columns and the corresponding Seurat object
        max_ncol <- current_ncol
        max_seurat <- seurat_rna
        max_index <- i  # Store the index of the maximum Seurat object
    }
}
#use the max index in the function
rna.anchors <- FindIntegrationAnchors(object.list = seurat_rna_list, anchor.features = features, reference=max_index)
saveRDS(rna.anchors, file = "Z:/dmclab/Marta/Combined_snRNAseq/data/rna.anchors.rds")

rna.combined <- IntegrateData(anchorset = rna.anchors)
saveRDS(rna.combined, file = "Z:/dmclab/Marta/Combined_snRNAseq/data/rna.combined_snRNAseq.rds")
```

```{r run-pca, eval=FALSE}

rna.combined <- RunPCA(object = rna.combined, assay = "SCT", features = features, npcs = 50, verbose = FALSE, reduction.name = "pca_before_harmony", seed.use = 17)
```

```{r elbowplot, fig.height=5, fig.width=6}

e<-ElbowPlot(object = rna.combined, ndims = 50, reduction = "pca_before_harmony") + scale_y_continuous(breaks = seq(0, 18, by = 2))+ggtitle("Elbow plot to select significant PCs")+ geom_hline(yintercept = 2, col="red", linetype= "dashed")
```

```{r eval=FALSE}
rna.combined <- RunUMAP(object = rna.combined, dims = 1:42, assay = "SCT", seed.use = 63, reduction = "pca_before_harmony", reduction.name = "umap_before_harmony") #38 contributing PCs chosen for further analysis
saveRDS(rna.combined, file = "Z:/dmclab/Marta/Combined_snRNAseq/secondoround/rna.combined_snRNAseq_integrated.rds") 
```

Control_4 \<- C1 Control_5 \<- C2 Mitopark_1 \<- MP1 Mitopark_2 \<- MP2 Mitopark_3 \<- MP3 6OHDA_1 \<- new_6OHDA_1 6OHDA_2 \<- new_6OHDA_2 6OHDA_3 \<- new_6OHDA_3 Control_1 \<- new_control_1 Control_2 \<- new_control_2 Control_3 \<- old_control_1


```{r UMAP_beforebatcheffects, message=FALSE, fig.width=16, fig.height=14, fig.show='hold'}
p1 <- DimPlot(object = rna.combined, group.by = "Sample", reduction = "umap_before_harmony") + ggtitle("Before batch correction (spots grouped by sample)")
be7b7eff
p2 <- DimPlot(object = rna.combined, group.by = "Sample_group",pt.size=1, reduction = "umap_before_harmony",cols=c("#84484aff", "gray","#d8b0b2ff", "#be7b7eff"), raster=FALSE)  + ggtitle("sample groups")

p3 <- FeaturePlot(object = rna.combined, features = "nCount_SCT", reduction = "umap_before_harmony")
p4 <- FeaturePlot(object = rna.combined, features = "nCount_RNA", reduction = "umap_before_harmony")

p <- DimPlot(object = rna.combined, 
             group.by="Sample_group",
              pt.size = 1, 
              reduction = "umap_before_harmony", 
              raster = FALSE) + 
  ggtitle("Sample groups") +
  scale_color_manual(values = c("#84484aff", "gray","#d8b0b2ff", "#be7b7eff"))

plot_filename <- paste0("Z:/dmclab/Marta/Combined_snRNAseq/new_allen_based_clustering/umap_before_harmony_cols_", ".pdf")
  #png(plot_filename, width = 600, height = 600) 
 pdf(plot_filename, width = 8, height = 6)  # Adjust width and height as desired
 print(p)
 dev.off()

(p1 - p2) / (p3 - p4)
```

###     Batch effects correction

```{r eval=FALSE}
library(harmony)
set.seed(57)
rna.combined <- RunHarmony(object = rna.combined, group.by.vars = c("Sample"), theta = c(1), assay.use = "SCT", reduction = "pca_before_harmony", dims.use = 1:42, plot_convergence = T, verbose = F, reduction.save = "harmony_sid_1")

rna.combined <- RunUMAP(object = rna.combined, assay.use = "SCT", reduction = "harmony_sid_1", dims = 1:42, seed.use = 6129, reduction.name = "umap_after_harmony")

set.seed(57)
rna.combined <- RunHarmony(object = rna.combined, group.by.vars = c("Sample_group"), theta = c(1), assay.use = "SCT", reduction = "pca_before_harmony", dims.use = 1:42, plot_convergence = T, verbose = F, reduction.save = "harmony_sid_2")

rna.combined <- RunUMAP(object = rna.combined, assay.use = "SCT", reduction = "harmony_sid_2", dims = 1:42, seed.use = 6129, reduction.name = "umap_after_harmony_2")

set.seed(57)
rna.combined <- RunHarmony(object = rna.combined, group.by.vars = c("Sample","Sample_group"), theta = c(1,1), assay.use = "SCT", reduction = "pca_before_harmony", dims.use = 1:42, plot_convergence = T, verbose = F, reduction.save = "harmony_sid_3")

rna.combined <- RunUMAP(object = rna.combined, assay.use = "SCT", reduction = "harmony_sid_3", dims = 1:42, seed.use = 6129, reduction.name = "umap_after_harmony_3")



```

```{r after_batcheffects, fig.width=16, fig.height=14, fig.show='hold'}
p1 <- DimPlot(object = rna.combined, group.by = "Sample", reduction = "umap_after_harmony") + ggtitle("After batch correction (spots grouped by sample)")

p2 <- DimPlot(object = rna.combined, group.by = "Sample_group", reduction = "umap_after_harmony", alpha=0.6, cols=c("yellow", "red","blue", "lightgreen"))  + ggtitle("sample groups")
p3 <- FeaturePlot(object = rna.combined, features = "nCount_SCT", reduction = "umap_after_harmony")
p4 <- FeaturePlot(object = rna.combined, features = "nCount_RNA", reduction = "umap_after_harmony")

(p1 - p2) / (p3 - p4)

```

```{r after_batcheffects, fig.width=16, fig.height=14, fig.show='hold'}
p1 <- DimPlot(object = rna.combined, group.by = "Sample", reduction = "umap_after_harmony_2") + ggtitle("After batch correction (spots grouped by sample group)")

p2 <- DimPlot(object = rna.combined, group.by = "Sample_group", reduction = "umap_after_harmony_2", alpha=0.6, cols=c("yellow", "red","blue", "lightgreen"))  + ggtitle("sample groups")

p3 <- FeaturePlot(object = rna.combined, features = "nCount_SCT", reduction = "umap_after_harmony_2")
p4 <- FeaturePlot(object = rna.combined, features = "nCount_RNA", reduction = "umap_after_harmony_2")

(p1 - p2) / (p3 - p4)
```

```{r after_batcheffects, fig.width=16, fig.height=14, fig.show='hold'}
p1 <- DimPlot(object = rna.combined, group.by = "Sample", reduction = "umap_after_harmony_3") + ggtitle("After batch correction (spots grouped by sample and sample group)")

p2 <- DimPlot(object = rna.combined, split.by = "Sample_group", reduction = "umap_after_harmony_3", alpha=0.6, cols=c("yellow", "red","blue", "lightgreen"))  + ggtitle("sample groups")

p3 <- FeaturePlot(object = rna.combined, features = "nCount_SCT", reduction = "umap_after_harmony_3")
p4 <- FeaturePlot(object = rna.combined, features = "nCount_RNA", reduction = "umap_after_harmony_3")

plot_filename <- paste0("Z:/dmclab/Marta/Combined_snRNAseq/stat_plots/umap_before_harmony_cols_split", ".pdf")
  #png(plot_filename, width = 600, height = 600) 
pdf(plot_filename, width = 8, height = 6)  # Adjust width and height as desired
print(p2)
dev.off()

(p1 - p2) / (p3 - p4)
```

# Clustering

Reduction selected: harmony_sid_3, umap: umap_after_harmony_3

PCs: 42

Resulution :0.4

```{r eval=FALSE}
#choose the reduction accondign to the one that fits best before running the next step, chosen umap_after_harmony_3
set.seed(61)
rna.combined <- FindNeighbors(object = rna.combined, assay = "SCT", dims = 1:42, k.param = 23, graph.name = "graph_afterHarmony", reduction = "harmony_sid_3")

rna.combined <- FindClusters(object = rna.combined, pc.use = 1:42, resolution = 0.4, save.SNN = T, do.sparse = T, graph.name = "graph_afterHarmony", random.seed = 13, group.singletons = TRUE)
saveRDS(rna.combined, file = "Z:/dmclab/Marta/Combined_snRNAseq/data/rna.combined.snRNA_seq_merged_42_res0.4_annotated_allen.rds")
rna.combined<- readRDS( file = "Z:/dmclab/Marta/Combined_snRNAseq/data/rna.combined.snRNA_seq_merged_42_res0.4_annotated_allen.rds")
```

```{r umap_clusters, fig.height=7, fig.width=18}
#cl.colors <-  c("#F18F01", "#878787", "#FCCDE5", "#048BA8", "#2E4057", "#99C24D", "#B95F89", "#5F070C" ,  "#DFC27D", "#DB6C79", "#B191FF", "#157A6E", "#EB4511", "#73683B", "#97C8EB","#C51B7D", "#BA9763", "#31081F", "#52D1DC", "#700353", "#F5D3C8", "#725752", "#D8315B", "#6B4E71", "#8CBFCF" , "#C2B2B4", "#690375" ,"#EDE5A6", "#52B788")
#pie(rep(1,length(cl.colors)), col = cl.colors)
#plot witht he correct reduction according to the batch correction selected

plot_directory <- "Z:/dmclab/Marta/Combined_snRNAseq/new_allen_based_clustering"


# Create the directory if it doesn't exist
dir.create(plot_directory, showWarnings = FALSE)

U1<-DimPlot(object = rna.combined, group.by = "seurat_clusters", reduction = "umap_after_harmony_3", pt.size = 0.6, label = T, label.size = 4, raster=TRUE) + ggtitle("Clusters in combined snRNAseq dataset RES 0.5")
plot_filename <- paste0(plot_directory, "/UMAP_41PC_RES0.4_", ".pdf")
  #png(plot_filename, width = 600, height = 600) 
pdf(plot_filename, width = 8, height = 8)  # Adjust width and height as desired
print(U1)
dev.off()

U2<-DimPlot(object = rna.combined, group.by = "seurat_clusters", reduction = "umap_after_harmony_3",split.by="Sample_group", pt.size = 0.6, label = T, label.size = 4 )
U3<-(U1 - U2)
plot_filename <- paste0(plot_directory, "/UMAP_40PC_RES0.4_SPLIT", ".pdf")
  #png(plot_filename, width = 600, height = 600) 
pdf(plot_filename, width = 24, height = 8)  # Adjust width and height as desired
print(U2)
dev.off()

grid.arrange(U1, U2, ncol = 2, widths = c(0.3, 0.7))

```

### Nuclei counts per cluster

```{r table1}
table<- as.data.frame(table("cluster" = rna.combined@active.ident, rna.combined@meta.data$orig.ident))
write.csv(table, file="Z:/dmclab/Marta/Combined_snRNAseq/new_allen_based_clustering/Nuclei_per_cluster_merged.csv")
table_plot <- ggplot(table, aes(x = cluster, y = Freq, fill = cl.colors)) +
  geom_col() +
  geom_text(aes(label = Freq), vjust = -0.5) +
  #xlim(0,3000)+
  labs(x = "counts", y = "Cluster") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        panel.grid = element_blank())
plot_filename <- paste0(plot_directory, "/CELLS_PER_CLUSTER_RES0.4", ".pdf")
  #png(plot_filename, width = 600, height = 600) 
pdf(plot_filename, width = 6, height =4)  # Adjust width and height as desired
print(table_plot)
dev.off()
```

### Nuclei counts per cluster across the different samples

```{r table2}
table1<-table("cluster" = rna.combined@active.ident, rna.combined@meta.data$Sample)

write.csv(table1, file="Z:/dmclab/Marta/Combined_snRNAseq/new_allen_based_clustering/Nuclei_across_samples_merged.csv")
```

### Nuclei counts per sample group

```{r table3}
table2<-table("cluster" = rna.combined@active.ident, rna.combined@meta.data$Sample_group)
write.csv(table2, file="Z:/dmclab/Marta/Combined_snRNAseq/new_allen_based_clustering/Nuclei_per_sample_group_merged.csv")
```

# Cluster Markers

### Top 5 markers for each cluster

-   Genes included only if they are expressed in atleast 10% of cells, hence not testing infrequently expressed genes within the two populations being compared. Here in this case, markers for cluster X mean cluster X vs all other clusters.
-   *adjusted p-value* threshold used here is 0.05
-   genes ordered also by *average log FC* values

```{r verbose=FALSE, warning=FALSE, message=FALSE, eval=FALSE}
DefaultAssay(rna.combined) <- "SCT"


rna.combined <- PrepSCTFindMarkers(object = rna.combined_all, assay = "SCT", verbose = TRUE)
wnn_markers <- FindAllMarkers(rna.combined_all, only.pos = TRUE, min.pct = 0.10, logfc.threshold = 0.25)
wnn_markers <- wnn_markers[wnn_markers$p_val_adj < 0.01, ]
write.csv(wnn_markers, file = "Z:/dmclab/Marta/Combined_snRNAseq/new_allen_based_clustering/cluster_snRNAseq_combined_42Pcs_res0.4_annotation_allen2.csv", col.names = TRUE, row.names = TRUE)

wnn_markers<- read.csv("Z:/dmclab/Marta/Combined_snRNAseq/new_allen_based_clustering/cluster_snRNAseq_combined_42Pcs_res0.4_annotation_allen_2.csv")
df_filtered <- wnn_markers[!grepl("^Gm|Rik$", wnn_markers$gene), ]
top5_2<-df_filtered %>%   
    group_by(cluster) %>%
    slice_max(n =10, order_by = avg_log2FC) %>%
    print(n = 1000)

genes_heatmap <- top5_2[,c(3,7,8)]

genes_heatmap$cluster <- factor(genes_heatmap$cluster, levels=clusters)

```

```{r warning=FALSE, message=FALSE, results='asis'}
#make sure that the markers are representative for the clusters, and that they are present in all conditions
DefaultAssay(rna.combined) <- "RNA"


# cons.markers <- FindConservedMarkers(rna.combined, ident.1 = , grouping.var = "Sample_group", verbose = FALSE)

# Create an empty dataset to store the results
cons_markers_dataset <- data.frame()

# Get unique values of the "userat_cluster" column
unique_ident_values <- unique(rna.combined$seurat_cluster)

# Loop through unique_ident_values
for (ident_value in unique_ident_values) {
  # Execute the function with the current ident_value
  cons_markers <- FindConservedMarkers(rna.combined, ident.1 = ident_value, grouping.var = "Sample_group", verbose = FALSE)
  
  # Add the ident_value to the result for reference
  cons_markers$ident_value <- ident_value
  
  # Append the current result to the dataset
  cons_markers_dataset <- rbind(cons_markers_dataset, cons_markers)
}

# View the resulting dataset
print(cons_markers_dataset)

```

```{r warning=FALSE, message=FALSE, results='asis'}


markers<-c("Mbp","Tnc","Cx3cr1","Slc1a3","Sox10","Pdgfra","Gfap","Aldh1l1","Slc17a7","Slc17a6" ,"Slc32a1","Gad2","Sst","Npy","Pvalb", "Chat", "Ache","Slc18a3","Gpr6","Ppp1r1b","Rgs9" ,"Gpr88","Adora2a","Drd2","Penk","Pdyn","Tac1","Drd1","Col11a1","Sema5b","Oprm1","Th")
d1_markers<- c("Gpr149", "Fam155a", "Syt6", "Neto2", "Rgs20", "Homer1", "Kcnj3", "Drd3", "Kremen1", "Egr1", "Lrrk2", "Tshz1", "Nr4a1")
st_markers<- c("Calb1","Cck","Ptgds","Enpp2", "Lars2","Cldn11","Arc", "Egr1","Avp","Col11a1","Tac1","Rgs2","Bc1","Pde7b","Otof","Nts")

all_genes<- c(d1_markers,st_markers,markers)

mark1<- c( "Nrxn1", "Nrxn3", "Dlg2", "Nlgn1","Grin3a", "Glra2", "Glra3",  "Gpr158")
mark_poster<- c("Drd1","Adora2a","Sema5b","Kremen1", "Tac1","Nr4a1","Rgs20","Rgs2","Homer1", "Syt6","Neto3")
```

```{r markers_dotplot, warning=FALSE, message=FALSE, verbose=FALSE, fig.width=18, fig.height= 10}


my_table<- as.data.frame(table("cluster" = rna.combined@active.ident, rna.combined@meta.data$orig.ident))


p <- DimPlot(object = rna.combined, group.by = "seurat_clusters", reduction = "umap_after_harmony_3",
             pt.size = 0.6, label = TRUE, label.size = 6)+ggtitle("Clustering of all Mitopark samples snRNA seq data")

# Create the table_plot and use the cluster colors
table_plot <- ggplot(my_table, aes(x = cluster, y = Freq, fill = cluster)) +
  geom_col() +

  labs(x = "counts", y = "Cluster") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        panel.grid = element_blank())

plot<- p+ table_plot
#plot<- grid.arrange(p, table_plot, ncol = 2)
plot_filename <- paste0(plot_directory, "/ALL_UMAP_42PC_RES0.4", ".pdf")
  #png(plot_filename, width = 600, height = 600) 
pdf(plot_filename, width = 16, height = 8)  # Adjust width and height as desired
print(plot)
dev.off()
top5_2 <- df_filtered %>%
  group_by(cluster) %>%
  slice_max(n = 5, order_by = avg_log2FC) %>%
  filter(row_number() %in% 1:5)

genes <- c("Mbp", "Prr5l", "Mog", "Inpp5d", "Cx3cr1", "Siglech", "Ly86", "Rgs9", "Gpr88", "Ppp1r1b", 
           "Drd1", "Tac1", "Pdyn", "Drd2", "Penk", "Adora2a", "Car10", "Tafa1", "Slc17a7", "Slc1a3", 
           "Aldh1l1", "Tnc", "Gfap", "Slco1c1", "Sst", "Npy", "Pvalb", "Kcnmb2", "Pdgfra")


# Define your custom order for the clusters
# Define your custom order for the clusters
custom_order <- c("oligodendrocytes", "Microglia","Astrocytes", "OPCs", "Cortex", "Striatum D1", "Striatum D2","Cholinergic Interneurons", "Interneurons")  
d3<-DotPlot(rna.combined_all, features =unique(top5_2$gene)) +
  scale_colour_gradientn(colours =rev(c("white","gray", "#B34B50", "#5F070C","#240705") )%>% rev()) + labs(y = "cluster")+
   theme(axis.text.x = element_text(angle = 90, vjust = 0, hjust=0))+
  scale_x_discrete(position = "bottom") +
  #scale_y_discrete(limits = rev(custom_order)) +
  theme(axis.text = element_text(family = "sans", size = 8),
        axis.title = element_text(family = "sans", size = 6),
        legend.key.size = unit(5, units = "mm"),
        legend.text = element_text(family = "sans", size = 5),
        legend.title = element_text(family = "sans", size = 6))+
 scale_size(range = c(1,8))+
  coord_flip()
plot_filename <- paste0(plot_directory, "/top5_markers_annotated2", ".pdf")
pdf(plot_filename, width = 6, height = 20)  # Adjust width and height as desired
print(d3)
dev.off()

```

### Cluster markers on UMAP

# Annotations using known markers list

#load dataset


```{r eval=FALSE}
#load combined
rna.combined <- readRDS("Z:/dmclab/Marta/Combined_snRNAseq/data/rna.combined.snRNA_seq_merged_42_res0.4_annotated_allen.rds")

wnn_markers<- read.csv("Z:/dmclab/Marta/Combined_snRNAseq/cluster_snRNAseq_combined_42Pcs_res0.4_annotated.csv")
  
  
  # Filter by significance (adjust based on your data)
filtered_markers <- wnn_markers[wnn_markers$p_val_adj < 0.001 & wnn_markers$avg_log2FC > 0.25, ]
marker_genes <- unique(filtered_markers$gene)

library(biomaRt)

# Connect to the Ensembl database for mouse
ensembl <- useMart("ensembl", dataset = "mmusculus_gene_ensembl")

# Map common gene names (MGI symbols) to Ensembl IDs
mapping <- getBM(
  filters = "mgi_symbol",
  attributes = c("mgi_symbol", "ensembl_gene_id"),
  values = marker_genes,
  mart = ensembl
)

# View the mapping results
head(mapping)

# Inspect features of the Seurat object
head(rownames(rna.combined))

# Find common genes between Seurat features and your mapping
common_genes <- intersect(rownames(rna.combined), mapping$mgi_symbol)
length(common_genes)  # Check how many genes overlap

# Subset the Seurat object based on common genes
seurat.obj <- subset(rna.combined, features = common_genes)



# Extract the normalized gene expression matrix (RNA assay)
expression_matrix <- GetAssayData(seurat.obj, assay = "RNA", slot = "data")
# Save the expression matrix
write.csv(as.matrix(expression_matrix), "Z:/dmclab/Marta/Combined_snRNAseq/expression_matrix.csv", row.names = TRUE)

# Save cell metadata (optional)
write.csv(data.frame(cell_names = colnames(expression_matrix)), "Z:/dmclab/Marta/Combined_snRNAseq/cell_metadata.csv", row.names = FALSE)

# Save gene metadata (optional)
write.csv(data.frame(gene_names = rownames(expression_matrix)), "Z:/dmclab/Marta/Combined_snRNAseq/gene_metadata.csv", row.names = FALSE)

print("Matrix and metadata saved as CSV.")


writeH5AD(anndata_obj, "Z:/dmclab/Marta/Combined_snRNAseq/filtered_expression_with_metadata.h5ad")

# Store sample-based metadata in a dataframe called obs
obs = rownames(count_matrix)
	 
# Store gene-based metadata in a dataframe called var2
var2 = colnames(count_matrix)


#load results 

# Step 1: Load the MapMyCells results
mapmycells_results <- read.csv("Z:/dmclab/Marta/Combined_snRNAseq/new_adata_for_MapMyCells_.csv", row.names = 1)

# Step 2: Check alignment of cell names
# Inspect Seurat object cell names
head(colnames(rna.combined))

# Inspect MapMyCells result cell names
head(rownames(mapmycells_results))

# Ensure alignment
if (!all(rownames(mapmycells_results) %in% colnames(rna.combined))) {
  stop("Cell names in MapMyCells results do not match Seurat object cell names.")
}

# Step 3: Merge MapMyCells annotations into Seurat metadata
rna.combined <- AddMetaData(
  object = rna.combined,
  metadata = mapmycells_results
)

# Inspect updated metadata
head(seurat.obj@meta.data)



# Access the metadata
metadata <- rna.combined@meta.data

# Remove the first 4 digits followed by a space and the trailing "_<number>"
metadata$cluster_name2 <- gsub("^[0-9]{4} |_[0-9]+$", "", metadata$cluster_name)
# Save the updated metadata back into the Seurat object
rna.combined@meta.data <- metadata

# Inspect the updated metadata
head(rna.combined@meta.data)


```

```{r dotplot_markers_manual, warning=FALSE, message=FALSE, verbose=FALSE, fig.width=12, fig.height= 10}
DefaultAssay(rna.combined) <- "SCT"
```


## Let's check UMAPs for these known markers

**Harmony**:

for 6OHDA: umap_after_harmony

for Mitopark: umap_after_harmony_3

for combined:umap_after_harmony_3

```{r umap_clusters2, fig.height=7, fig.width=18}
#pie(rep(1,length(cl.colors)), col = cl.colors)


rna.combined<- readRDS("Z:/dmclab/Marta/Combined_snRNAseq/data/rna.combined.snRNA_seq_merged_42_res1.5_annotated_allen.rds")
p<-DimPlot(object = rna.combined, group.by = "cluster_name2", reduction = "umap_after_harmony_3", pt.size = 0.6, label = T, label.size = 6) + ggtitle("Clusters in combined snRNAseq dataset")
#check that the levels are actually in the order or Idents
Idents(rna.combined)
#annotation for combined dataset


p <-  DimPlot(
  object = rna.combined,
    group.by = "cluster_name",
  reduction = "umap_after_harmony_3",
  pt.size = 0.6,
  label = TRUE,
  label.size = 3
) + 
  ggtitle("Clusters in combined snRNAseq dataset") +
  theme(legend.position = "none")       # Adjust legend key size

# Determine the number of clusters
n_clusters <- length(unique(rna.combined_all$class_name))

library(RColorBrewer)
# Get the "Paired" palette from RColorBrewer
palette <- brewer.pal(n = min(n_clusters, 12), name = "Paired")  # Adjust for max palette size

# Extend the palette if more than 12 clusters (optional)
if (n_clusters > 12) {
  palette <- colorRampPalette(brewer.pal(12, "Paired"))(n_clusters)
}

# Apply the palette to DimPlot
p <- DimPlot(
  object = rna.combined_all,
  group.by = "class_name",
  reduction = "umap_after_harmony_3",
  pt.size = 0.6,
  label = TRUE,
  label.size = 2
) + 
  scale_color_manual(values = palette)
plot_filename <- paste0(plot_directory, "/UMAP_ANNOTATED_pc42_res0.4_SIZED_class", ".pdf")
  #png(plot_filename, width = 1800, height = 600) 
  pdf(plot_filename, width = 11, height = 8)  # Adjust width and height as desired
  print(p)
  dev.off()


#delete old annotation
Idents(rna.combined_all) <- rna.combined_all@meta.data$seurat_clusters
new.cluster.ids<- c("oligodendrocytes","Microglia","oligodendrocytes","Striatum D1", "Cortex IT−ET","oligodendrocytes","Striatum D2","Splatter","Cortex CT−L6b","oligodendrocytes","Astrocytes/Epen","Chat/Npy/Pval+ Interneurons","OPCs","eSPN","Cortex IT−ET","Interneurons","Vascular","Interneurons","oligodendrocytes","Cortex IT−ET","Cortex CT−L6b","oligodendrocytes","oligodendrocytes")

names(new.cluster.ids) <- levels(rna.combined_all)
rna.combined_all <- RenameIdents(rna.combined_all, new.cluster.ids)
rna.combined_all$clusters <- Idents(rna.combined_all)

cl.colors <-  c("#F5D3C8", "#558877ff","#0a4347ff","#b89a01ff","#B95F89","#484444ff","#dace7eff","#977482ff","#061c8dff","#FCCDE5","#856fa7ff","#102c52","#5F070C")

cl.colors2 <-  c("#F5D3C8", "#878787","#0a4347ff","#878787","#878787","#878787","#878787","#878787","#878787","#878787","#878787","#878787")

saveRDS(rna.combined_all, "Z:/dmclab/Marta/Combined_snRNAseq/data/rna.combined.snRNA_seq_merged_42_res0.4_annotated_allen.rds")
# Set a directory to save the plots
plot_directory <- "Z:/dmclab/Marta/Combined_snRNAseq/new_allen_based_clustering"


# Create the directory if it doesn't exist
dir.create(plot_directory, showWarnings = FALSE)

p2<-DimPlot(object = rna.combined,  reduction = "umap_after_harmony_3",pt.size = 0.3, label = F,split.by="Sample_group", label.size = 6,  raster=FALSE ,cols=alpha(cl.colors,0.5) ) + ggtitle("Clusters in combined snRNAseq dataset") & theme_void()
plot_filename <- paste0(plot_directory, "/UMAP_ANNOTATED_pc42_res0.4_SIZED_ann_split", ".pdf")
  #png(plot_filename, width = 1800, height = 600) 
  pdf(plot_filename, width = 44, height = 8)  # Adjust width and height as desired
  print(p2)
  dev.off()

  my_cols<- c("#858585", "#EE956C")
  my_cols<- c("#ffd694ff","#858585", "#96c1b2ff","#4a7f6dff")
p2<-DimPlot(object = rna.combined,  reduction = "umap_after_harmony_3",pt.size = 0.5, label = F, label.size = 6,group.by="Sample_group",raster=FALSE, cols=my_cols  ) + ggtitle("Clusters in combined snRNAseq dataset") & theme_void()
plot_filename <- paste0(plot_directory, "/UMAP_ANNOTATED_pc42_res0.4_group_cols_AFFETCED", ".pdf")
  #png(plot_filename, width = 1800, height = 600) 
  pdf(plot_filename, width = 8, height = 6)  # Adjust width and height as desired
  print(p2)
  dev.off()  
  
  
  #calculate average number of genes per cluster

# Step 1: Extract the cluster information and gene expression matrix
clusters <- Idents(rna.combined)            # Cluster identities for each cell
gene_expression <- rna.combined@assays$RNA@data  # Expression data matrix

# Step 2: Calculate the number of genes expressed per cell (non-zero count)
genes_per_cell <- colSums(gene_expression > 0)

# Step 3: Calculate the average number of genes per cluster
avg_genes_per_cluster <- tapply(genes_per_cell, clusters, mean)

# View results
mean(avg_genes_per_cluster)
  
rna.combined_all@meta.data$orig.ident <- "snRNA seq experiments"
table<- table("cluster" = rna.combined_all@active.ident, rna.combined_all@meta.data$orig.ident, rna.combined@meta.data$Sample_group)
table<- as.data.frame(table)

# Calculate percentages
Percentage <- (table$Freq / sum(table$Freq)) * 100
table1<- cbind(table[1],table[3],table[4],Percentage, cl.colors)

Sample_groups<- unique(table$Var3)
table_final<- data.frame()
for (i in Sample_groups){
  table_group<- subset(table, Var3== i)
  Percentage <- (table_group$Freq / sum(table_group$Freq)) * 100
  table1<- cbind(table_group[1],table_group[3],table_group[4],Percentage, cl.colors)
  table_final<- rbind(table1, table_final)

}



p <- VlnPlot(object = rna.combined_all, features = "nFeature_RNA",  pt.size = 0, assay = "RNA", cols= cl.colors) 
ggsave(p, filename = paste0(plot_directory, "/violins_nGene_cluster.pdf"), dpi = 300, width = 10, height = 5)

colnames(table_final)<- c("cluster","Sample_group", "Freq", "Percentage", "color")

p<-ggplot(table_final, aes(x = Sample_group, y = Percentage, fill = cluster)) +
  geom_bar(stat = "identity", width = 0.6) +
  labs(x = "Sample Group", y = "Percentage") +
  scale_fill_manual(values = table_final$color) +  # Assign colors
  theme_classic() +  # Set plot theme
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) 
plot_filename <- paste0(plot_directory, "/proportions_stacked", ".pdf")
  #png(plot_filename, width = 1800, height = 600) 
  pdf(plot_filename, width = 5, height = 5)  # Adjust width and height as desired
  print(p)
  dev.off()  
  
# Create a pie chart using ggplot2 with custom colors and labels
pie_chart <- ggplot(table1, aes(x = "", y = Percentage, fill = cluster)) +
  geom_bar(stat = "identity", width = 1) +
  coord_polar(theta = "y") +
  theme_void() +
  scale_fill_manual(values=cl.colors)+
  ggtitle("Cluster Distribution") +
  geom_text(aes(label = paste0(round(Percentage, 2), "%")), position = position_stack(vjust = 0.5)) 
plot_filename <- paste0(plot_directory, "/proportions_pie", ".pdf")
  #png(plot_filename, width = 1800, height = 600) 
  pdf(plot_filename, width = 5, height = 5)  # Adjust width and height as desired
  print(pie_chart)
  dev.off()  
  


rna.combined <- ValidateClusters(rna.combined, pc.use = 1:42, top.genes = 60)

pbmc <- BuildClusterTree(pbmc, 
                         do.reorder = T, 
                         reorder.numeric = T)

combined_averages <- AverageExpression(controls, return.seurat = TRUE) 
my_levels <- c(7, 6, 4, 3, 1, 2, 5, 10, 8, 0, 9, 11, 12)

# Re-level object@ident
yourseuratobject@active.ident <- factor(x = yourseuratobject@active.ident, levels = my_levels)

my_levels<- c("Astrocytes","Cholinergic Interneurons", "Cortex","OPCs","Interneurons","Microglia","Striatum D1","Striatum D2","oligodendrocytes")
levels(combined_averages) <- my_levels
library(Seurat)
HEATMAP<- DoHeatmap(
  combined_averages,
  features =unique(genes_to_find),
  cells =NULL ,
  group.by = "ident",
  group.bar = TRUE,
  group.colors = cl.colors,
  disp.min = -2.5,
  disp.max = NULL,
  slot="data",
  assay="SCT",
  label = TRUE,
  size = 3,
  hjust = 0,
  vjust = 0,
  angle = 45,
  raster = FALSE,
  draw.lines = FALSE,
  lines.width = NULL,
  group.bar.height = 0.02,
  combine = TRUE
) + scale_fill_gradientn(colors = c("white","#B34B50", "#5F070C","#240705"))

HEATMAP<-HEATMAP + theme(
  axis.text = element_text(size = 12),  # Change the size of tick labels
  axis.title = element_text(size = 6)  # Change the size of axis titles
)
plot_filename <- paste0(plot_directory, "/heatmap_ctrl_transmitters_all", ".pdf")
  #png(plot_filename, width = 600, height = 600) 
  pdf(plot_filename, width = 8, height = 15)  # Adjust width and height as desired
  print(HEATMAP)
  dev.off()

markers<-c("Mbp","Tnc","Cx3cr1","Slc1a3","Sox10","Pdgfra","Gfap","Aldh1l1","Slc17a7","Slc17a6" ,"Slc32a1","Gad2","Sst","Npy","Pvalb","Chat","Slc18a3","Gpr6","Ppp1r1b","Rgs9" ,"Gpr88","Crym","Gpr155","Adora2a","Drd2","Penk","Pdyn","Tac1","Drd1","Col11a1","Sema5b","Oprm1")
levels=c("Mbp","Tnc","Cx3cr1","Slc1a3","Sox10","Pdgfra","Gfap","Aldh1l1","Slc17a7","Slc17a6" ,"Slc32a1","Gad2","Sst","Npy","Pvalb","Chat","Ache","Slc18a3","Gpr6","Ppp1r1b","Rgs9" ,"Gpr88","Adora2a","Drd2","Penk","Pdyn","Tac1","Drd1","Col11a1","Sema5b","Oprm1","Th")

markers_1<- c("Mbp", "Atp13a5", "Slco1c1",  "Cx3cr1", "Kcnmb2", "Rgs9", "Gpr88", "Ppp1r1b", "Phactr1", "Drd1", "Tac1", "Pdyn", "Tafa1", "Car10", "Penk", "Drd2", "Adora2a", "Gpr6", "Slc17a7", "Slc1a3", "Aldh1l1", "Siglech", "Inpp5d", "Ly86", "Tnc", "Gfap", "Sst", "Npy", "Pvalb", "Pdgfra",  "Prr5l", "Mog")
levels<- c("Mbp", "Atp13a5", "Slco1c1",  "Cx3cr1", "Kcnmb2", "Rgs9", "Gpr88", "Ppp1r1b", "Phactr1", "Drd1", "Tac1", "Pdyn", "Tafa1", "Car10", "Penk", "Drd2", "Adora2a", "Gpr6", "Slc17a7", "Slc1a3", "Aldh1l1", "Siglech", "Inpp5d", "Ly86", "Tnc", "Gfap", "Sst", "Npy", "Pvalb", "Pdgfra",  "Prr5l", "Mog")

clusters<- unique(Idents(rna.combined))

# Define your custom order for the clusters
custom_order <- c("oligodendrocytes", "Microglia","Astrocytes", "OPCs", "Cortex", "Striatum D1", "Striatum D2","Cholinergic Interneurons", "Interneurons")  
d3<-DotPlot(rna.combined_all, features =markers) +
  scale_colour_gradientn(colours =rev(c("white","gray", "#B34B50", "#5F070C","#240705") )%>% rev()) + labs(y = "cluster")+
   theme(axis.text.x = element_text(angle = 90, vjust = 0.8, hjust=-0.3))+
  scale_x_discrete(position = "bottom") +
  #scale_y_discrete(limits = rev(custom_order))+
  theme(axis.text = element_text(family = "sans", size = 7),
        axis.title = element_text(family = "sans", size = 10),
        legend.key.size = unit(5, units = "mm"),
        legend.text = element_text(family = "sans", size = 7),
        legend.title = element_text(family = "sans", size = 10))+
 scale_size(range = c(1,8))
plot_filename <- paste0(plot_directory, "/selected_markers_annotated", ".pdf")
  #png(plot_filename, width = 600, height = 600) 
pdf(plot_filename,width=10, height = 4) # Adjust width and height as desired
print(d3)
dev.off()

```

```{r markers_manual_featureplots, fig.width=20, fig.height=7, fig.show='hold', warning=FALSE}
DefaultAssay(rna.combined) <- "SCT"

# Set a directory to save the plots
plot_directory <- "Z:/dmclab/Marta/Combined_snRNAseq/final_plots2"


# Create the directory if it doesn't exist
dir.create(plot_directory, showWarnings = FALSE)

lapply(unique(markers), function(gene){
                          
   myplot <- VlnPlot(object = rna.combined_all, features = gene, pt.size = 0,log=F,  cols = cl.colors)+
     ylim(0,2)

  plot_filename <- paste0(plot_directory, "/violin_axis_all", gene, ".pdf")
  pdf(plot_filename, width = 8, height = 6)  # Adjust width and height as desired
  print(myplot)
  dev.off()
})



```

```{r markers_featureplots, fig.width=13, fig.height=7, fig.show='hold', warning=FALSE}

mark<- c("Gm42418","Kcnip4","Sgcd","Cntnap5b","Gm32647","Ptchd4","Ntng1","Dpp10" )
mark2<- c("Astn2","Gpc6","Pde4d","Prkg1","Cntnap5b","Dpp10","Gm42418")

markers<- c("Nr4a1", "Tac1", "Rgs2","Rgs20","Adora2a", "Ptgds")

mrk<- c("Tac2","Nxph4","Crtac1","Rreb1","Sgcd","Ntng2",
"Ryr1","Stac")

mrk_fig<- c("Crym","Gpr155", "Dlk1","Tac2","Nxph4","Crtac1","Rreb1","Sgcd","Ntng2","Ryr1","Stac")
lapply(unique(vars), function(gene){
  tryCatch({
    myplot <- FeaturePlot(rna.combined, features = gene, reduction = "umap_after_harmony_3", slot = "data", pt.size=0.6,
                          label=FALSE, raster=FALSE) & 
                          theme(legend.position = "right") &
                          #scale_color_gradientn(colors = c("gray","#60795bff", "#354332ff"), limits=c(0,3))
                          scale_color_gradientn(colors = c("gray","#B34B50", "#5F070C","#240705"), limits=c(0,4))
      

    plot_filename <- paste0(plot_directory, "/marker_umap_HD_norm_fig_sized_", gene, ".pdf")
    pdf(plot_filename, width = 11, height = 8)  # Adjust width and height as desired
    print(myplot)
    dev.off()
  }, error = function(e) {
    message(paste("Error with gene:", gene, " - Skipping to next..."))
    # Continue with the next gene if there's an error
  })
)
 #      
 # myplot1 <-FeaturePlot(rna.combined, features = gene, reduction = "umap_after_harmony", slot = "data",label=TRUE )+ggtitle("Subclustering")& theme(legend.position = "right") 
 # 
 # # Save the plot
 # # Save the plot as PDF
 #  plot_filename <- paste0(plot_directory, "/", gene, "_sub_marker_norm_grey.pdf")
 #  pdf(plot_filename, width = 8, height = 6)  # Adjust width and height as desired
 #  print(myplot1)
 #  dev.off()


```

  #   DE genes for mitopark KO vs wt per cluster

file: mitopark-KO_wt_DEGs-cluster_2023-09-05.xlsx

-   Dotplots show DE genes filtered by adjusted p-value \< 0.01
-   Top genes are the top (upto 10 shown) upregulated genes for mitopark KO samples vs controls in each cluster


```{r message=FALSE, warning=FALSE}

#other method to check the DE with MAST instead of wilcox
options(future.globals.maxSize = 100 * 1024^3)  # 100 GiB 
de_genes_6ohda <- list()
de_genes_mp<- list()
de_genes_mp11<- list()

rna.combined_all<-rna.combined
rna.combined <- subset(x = rna.combined_all, idents = setdiff(unique(Idents(rna.combined_all)), "Splatter"))
DefaultAssay(rna.combined) <- "RNA"
rna.combined <- NormalizeData(rna.combined)
use_clusters <- rna.combined@active.ident
all_clusters <- as.character(sort(levels(unique(use_clusters))))



options(digits=15)
nuclei_count <-as.data.frame(table("cluster" = rna.combined@active.ident, rna.combined@meta.data$Sample_group))

#check the total number of cells in cluster and each sample




set.seed(57)
cl_names <- c()
for (cl in 1:length(all_clusters)) {
    print(all_clusters[cl])
    Ctrl_cells <- colnames(rna.combined)[rna.combined@active.ident %in% all_clusters[cl] & rna.combined$Sample_group %in% "Control"]
    tr_cells <- colnames(rna.combined)[rna.combined@active.ident %in% all_clusters[cl] & rna.combined$Sample_group %in% "6OHDA"]

     if (length(tr_cells) < 3 | length(Ctrl_cells) < 3){
        print("test")
        next
    }

    length_ct<- length(Ctrl_cells)
    length_tr<- length(tr_cells)
    
    if (length_ct > length_tr){
      
    num_cells_to_select <- length(tr_cells)
    selected_ct_cells <- sample(Ctrl_cells,num_cells_to_select)
    # 
    de_genes_6ohda[[cl]] <- FindMarkers(rna.combined[["RNA"]], slot = "data", cells.1 = tr_cells, cells.2 = selected_ct_cells, test.use = "MAST", latent.vars = "Sample", min.pct = 0.05, logfc.threshold = 0.2,            features = NULL)
      
    }else{
    
      #Randomly select the same number of cells from shuffled bigger group so to have the same number of cells in the conditions
    num_cells_to_select <- length(Ctrl_cells)
    selected_tr_cells <- sample(tr_cells,num_cells_to_select)
    # 
    de_genes_6ohda[[cl]] <- FindMarkers(rna.combined[["RNA"]], slot = "data", cells.1 = selected_tr_cells, cells.2 = Ctrl_cells, test.use = "MAST", latent.vars = "Sample", min.pct = 0.05, logfc.threshold = 0.2,      features = NULL)
    }
    
    de_genes_6ohda[[cl]]$Gene <- rownames(de_genes_6ohda[[cl]])
    
    cl_names <- union(cl_names, all_clusters[cl])
}
names(de_genes_6ohda) <- cl_names

library(openxlsx)
write.xlsx(de_genes_6ohda, file = "Z:/dmclab/Marta/Combined_snRNAseq/new_allen_based_clustering/Combined-6OHDA_wt_DEGs-cluster_res05_nospl.xlsx")
# Specify the file path of your Excel workbook
install.packages("readxl")
library(readxl)
excel_file <- "Z:/dmclab/Marta/Combined_snRNAseq/new_allen_based_clustering/Combined-6OHDA_wt_DEGs-cluster_res05_nospl.xlsx"

# Get the names of all sheets in the workbook
sheet_names <- excel_sheets(excel_file)

# Create an empty list to store data frames
de_genes_6ohda <- list()

# Loop through each sheet and read it into a data frame
for (sheet_name in sheet_names) {
  sheet_data <- read_excel(excel_file, sheet = sheet_name)
  de_genes_6ohda[[sheet_name]] <- sheet_data
}



cl_names <- c()
for (cl in 1:length(all_clusters)) {
    print(all_clusters[cl])
    Ctrl_cells <- colnames(rna.combined)[rna.combined@active.ident %in% all_clusters[cl] & rna.combined$Sample_group %in% "Control"]
    tr_cells <- colnames(rna.combined)[rna.combined@active.ident %in% all_clusters[cl] & rna.combined$Sample_group %in% "Mitopark15_18"]

     if (length(tr_cells) < 3 | length(Ctrl_cells) < 3){
        print("test")
        next
    }

    length_ct<- length(Ctrl_cells)
    length_tr<- length(tr_cells)
    
    if (length_ct > length_tr){
      
    num_cells_to_select <- length(tr_cells)
    selected_ct_cells <- sample(Ctrl_cells,num_cells_to_select)
    # 
    de_genes_mp[[cl]] <- FindMarkers(rna.combined[["RNA"]], slot = "data", cells.1 = tr_cells, cells.2 = selected_ct_cells, test.use = "MAST", latent.vars = "Sample", min.pct = 0.05, logfc.threshold = 0.2,            features = NULL)
      
    }else{
    
      #Randomly select the same number of cells from shuffled bigger group so to have the same number of cells in the conditions
    num_cells_to_select <- length(Ctrl_cells)
    selected_tr_cells <- sample(tr_cells,num_cells_to_select)
    # 
    de_genes_mp[[cl]] <- FindMarkers(rna.combined[["RNA"]], slot = "data", cells.1 = selected_tr_cells, cells.2 = Ctrl_cells, test.use = "MAST", latent.vars = "Sample", min.pct = 0.05, logfc.threshold = 0.2,      features = NULL)
    }
    
    de_genes_mp[[cl]]$Gene <- rownames(de_genes_mp[[cl]])
    
    cl_names <- union(cl_names, all_clusters[cl])
}

names(de_genes_mp) <- cl_names
write.xlsx(de_genes_mp, file = "Z:/dmclab/Marta/Combined_snRNAseq/new_allen_based_clustering/Combined-mp15_18_wt_DEGs-cluster_res05_nospl.xlsx")
excel_file <- "Z:/dmclab/Marta/Combined_snRNAseq/new_allen_based_clustering/Combined-mp15_18_wt_DEGs-cluster_res05_nospl.xlsx"
# Get the names of all sheets in the workbook
sheet_names <- excel_sheets(excel_file)

# Create an empty list to store data frames
de_genes_mp <- list()

# Loop through each sheet and read it into a data frame
for (sheet_name in sheet_names) {
  sheet_data <- read_excel(excel_file, sheet = sheet_name)
  de_genes_mp[[sheet_name]] <- sheet_data
}


cl_names <- c()
for (cl in 1:length(all_clusters)) {
    print(all_clusters[cl])
    Ctrl_cells <- colnames(rna.combined)[rna.combined@active.ident %in% all_clusters[cl] & rna.combined$Sample_group %in% "Control"]
    tr_cells <- colnames(rna.combined)[rna.combined@active.ident %in% all_clusters[cl] & rna.combined$Sample_group %in% "Mitopark_11"]

     if (length(tr_cells) < 3 | length(Ctrl_cells) < 3){
        print("test")
        next
    }

    length_ct<- length(Ctrl_cells)
    length_tr<- length(tr_cells)
    
    if (length_ct > length_tr){
      
    num_cells_to_select <- length(tr_cells)
    selected_ct_cells <- sample(Ctrl_cells,num_cells_to_select)
    # 
    de_genes_mp11[[cl]] <- FindMarkers(rna.combined[["RNA"]], slot = "data", cells.1 = tr_cells, cells.2 = selected_ct_cells, test.use = "MAST", latent.vars = "Sample", min.pct = 0.05, logfc.threshold = 0.2,            features = NULL)
      
    }else{
    
      #Randomly select the same number of cells from shuffled bigger group so to have the same number of cells in the conditions
    num_cells_to_select <- length(Ctrl_cells)
    selected_braak_cells <- sample(tr_cells,num_cells_to_select)
    # 
    de_genes_mp11[[cl]] <- FindMarkers(rna.combined[["RNA"]], slot = "data", cells.1 = selected_tr_cells, cells.2 = Ctrl_cells, test.use = "MAST", latent.vars = "Sample", min.pct = 0.05, logfc.threshold = 0.2,      features = NULL)
    }
    
    de_genes_mp11[[cl]]$Gene <- rownames(de_genes_mp11[[cl]])
    
    cl_names <- union(cl_names, all_clusters[cl])
}
names(de_genes_mp11) <- cl_names
write.xlsx(de_genes_mp11, file = "Z:/dmclab/Marta/Combined_snRNAseq/subclustering/new_allen_based_clustering/Combined-mp11_wt_DEGs-cluster_res05_nospl.xlsx")

excel_file <- "Z:/dmclab/Marta/Combined_snRNAseq/new_allen_based_clustering/Combined-mp11_wt_DEGs-cluster_res04.xlsx"
# Get the names of all sheets in the workbook
sheet_names <- excel_sheets(excel_file)

# Create an empty list to store data frames
de_genes_mp11 <- list()

# Loop through each sheet and read it into a data frame
for (sheet_name in sheet_names) {
  sheet_data <- read_excel(excel_file, sheet = sheet_name)
  de_genes_mp11[[sheet_name]] <- sheet_data
}


```

```{r message=TRUE, warning=FALSE}
#volcano plots woith DE genes


# Set a directory to save the plots
plot_directory <- "Z:/dmclab/Marta/Combined_snRNAseq/new_allen_based_clustering"


# Create the directory if it doesn't exist

dir.create(plot_directory, showWarnings = FALSE)
# de.d1<- subset(de.genes.treated, column_label=="Striatum D1")
# subset_data <- de.15.cl2[de.15.cl2$avg_log2FC >= -0.7 & de.15.cl2$avg_log2FC <= 0.7,]
#



de_6ohda<- bind_rows(de_genes_6ohda, .id = "cluster")
de_6ohda <- de_6ohda %>% filter(p_val_adj < 0.0001)
de_60hda_up<- de_6ohda %>% filter(avg_log2FC > 0.2)
de_60hda_down<- de_6ohda %>% filter(avg_log2FC < -0.2)
de_6ohda<- rbind(de_60hda_up,de_60hda_down)
# three clusters
table(rna.combined@active.ident)
cluster.set <- unique(rna.combined$nFeature_RNA)
# overall nFeature
sum(rowSums(rna.combined[['RNA']]@counts) != 0)


clusters<- unique(rna.combined@active.ident)
row_6ohda<- data.frame()

de_mp<- bind_rows(de_genes_mp, .id = "cluster")
de_mp <- de_mp %>% filter(p_val_adj < 0.0001)
de_mp_up<- de_mp %>% filter(avg_log2FC > 0.2)
de_mp_down<- de_mp %>% filter(avg_log2FC < -0.2)
de_mp<-rbind(de_mp_up, de_mp_down)

nuclei_countmp<-nuclei_count  %>% filter(Var2 %in% c("Control", "Mitopark15_18"))
 de_mp1<-de_mp %>% group_by(cluster) %>%
  summarise(count = n())

 de_mp1 <- de_mp %>%
  group_by(cluster) %>%
  summarise(count = n()) %>%
  mutate(proportion = count / sum(count))

de_mp11<- bind_rows(de_genes_mp11, .id = "cluster")
de_mp11 <- de_mp11 %>% filter(p_val_adj < 0.0001)
de_mp11_up<- de_mp11 %>% filter(avg_log2FC > 0.2)
de_mp11_down<- de_mp11 %>% filter(avg_log2FC < -0.2)
de_mp11<-rbind(de_mp11_up, de_mp11_down)

nuclei_countmp11<-nuclei_count  %>% filter(Var2 == "Mitopark_11")
 de_mp11_1<-de_mp11 %>% group_by(cluster) %>%
  summarise(count = n())

  de_mp11_1 <- de_mp11 %>%
  group_by(cluster) %>%
  summarise(count = n()) %>%
  mutate(proportion =(count / sum(count))*100)

clusters<- unique(Idents(rna.combined))
table_gene_count<- data.frame()
 for (clust in clusters){
   subset<- subset(rna.combined, Sample_group %in% c("Control", "6OHDA"))
   subset_cluster<- subset(subset,  idents =clust)
   subset_cluster <- subset_cluster[rowSums(subset_cluster) > 0, ]
   gene_count<- length(rownames(subset_cluster))
   cluster<- clust
   gene_counts_df<- as.data.frame(gene_count, cluster)
   table_gene_count<- rbind(table_gene_count,gene_counts_df )
 }

table_gene_count_mp<- data.frame()
 for (clust in clusters){
   subset<- subset(rna.combined, Sample_group  %in% c("Control","Mitopark15_18"))
   subset_cluster<- subset(subset,  idents =clust)
   subset_cluster <- subset_cluster[rowSums(subset_cluster) > 0, ]
   gene_count<- length(rownames(subset_cluster))
   cluster<- clust
   gene_counts_df<- as.data.frame(gene_count, cluster)
   table_gene_count_mp<- rbind(table_gene_count_mp,gene_counts_df )
   
 }


table_gene_count_mp11<- data.frame()
 for (clust in clusters){
   subset<- subset(rna.combined, Sample_group %in% c("Control", "Mitopark_11"))
   subset_cluster<- subset(subset,  idents =clust)
   subset_cluster <- subset_cluster[rowSums(subset_cluster) > 0, ]
   gene_count<- length(rownames(subset_cluster))
   cluster<- clust
   gene_counts_df<- as.data.frame(gene_count, cluster)
   table_gene_count_mp11<- rbind(table_gene_count_mp11,gene_counts_df )
   
 }
table_gene_count$Experiment<- "6OHDA"
cluster<- rownames(table_gene_count)
table_gene_count<- cbind(cluster, table_gene_count)
colnames(table_gene_count)<- c("cluster","total_genes","Experiment")
de_6ohda1<-de_6ohda %>% group_by(cluster) %>%
  summarise(count = n())
de_6ohda_merge <- merge(de_6ohda1, table_gene_count, by = "cluster")
de_6ohda_merge<-transform(de_6ohda_merge, norm = count / total_genes)
#de_6ohda_merge <- merge(de_6ohda_merge, table_colors, by = "cluster")

table_gene_count_mp$Experiment<- "Mitopark 15/18"
cluster<- rownames(table_gene_count_mp)
table_gene_count_mp<- cbind(cluster, table_gene_count_mp)
colnames(table_gene_count_mp)<- c("cluster","total_genes","Experiment")
de_mp1<-de_mp %>% group_by(cluster) %>%
  summarise(count = n())
de_mp_merge <- merge(de_mp1, table_gene_count_mp, by = "cluster")
de_mp_merge<-transform(de_mp_merge, norm = count / total_genes)
#de_mp_merge <- merge(de_mp_merge, table_colors, by = "cluster")


table_gene_count_mp11$Experiment<- "Mitopark 11"
cluster<- rownames(table_gene_count_mp11)
table_gene_count_mp11<- cbind(cluster, table_gene_count_mp11)
colnames(table_gene_count_mp11)<- c("cluster","total_genes","Experiment")
de_mp11_1<-de_mp11 %>% group_by(cluster) %>%
  summarise(count = n())
de_mp11_merge <- merge(de_mp11_1, table_gene_count_mp11, by = "cluster")
de_mp11_merge<-transform(de_mp11_merge, norm = count / total_genes)
#de_mp_merge <- merge(de_mp_merge, table_colors, by = "cluster")


total_genes<- rbind(de_6ohda_merge, de_mp_merge)


# Create the bar plot
p<- ggplot(total_genes, aes(x = reorder(cluster, norm), y = norm, fill = Experiment)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(x = "Clusters", y = "Significat changed genes/ total genes in the cluster (p value<0.0001)", title = "Genes affected per group") +
  #ylim(0,0.1)+
  scale_fill_discrete(name = "Cluster")+
   facet_grid(.~Experiment)+
  #theme(axis.text.x = element_text(angle = 45))+
    coord_flip()+
  theme_classic()
plot_filename <- paste0(plot_directory, "/barplot_de_mp11_norm_thresh02_allgenes", ".pdf")
  #png(plot_filename, width = 600, height = 600) 
  pdf(plot_filename, width = 13, height = 6)  # Adjust width and height as desired
  print(p)
  dev.off()

 # Step 2: Reorder by 'norm' in descending order
de_6ohda_merge<- de_6ohda_merge[order(de_6ohda_merge$norm, decreasing = TRUE), ]


RColorBrewer::display.brewer.pal(n=10, name = "Blues")
blues <- RColorBrewer::brewer.pal(n = 5, name = "Blues") 
# Step 3: Reverse the 'blues' palette
de_6ohda_merge <- de_6ohda_merge[order(de_6ohda_merge$norm, decreasing = FALSE), ] #reorder by affected clusters so that the last ones are the affected ones. We assign them the blues gradient in line below
de_6ohda_merge$colors <- c(rep("grey", 7), blues)


cl.colors.affected <- c("#08306bff","#c6dbefff","#07519cff","#6baed6ff","#3c8cc3ff","grey","grey","grey","grey","grey","grey")  #manually select colors ordered like the clusters in the umap

p <- DimPlot(object = rna.combined, reduction = "umap_after_harmony_3", pt.size = 0.6, label = F, cols =alpha(cl.colors.affected,1) ,raster=FALSE, label.size = 12) + ggtitle("Most affected clusters from mild 6ohda dataset")

plot_filename <- paste0(plot_directory, "/umap_de_6ohda_norm", ".pdf")
  #png(plot_filename, width = 600, height = 600) 
  pdf(plot_filename, width = 11, height = 8)  # Adjust width and height as desired
  print(p)
  dev.off() 
  
# Step 3: Reverse the 'blues' palette
de_mp_merge <- de_mp_merge[order(de_mp_merge$norm, decreasing = FALSE), ] #reorder by affected clusters so that the last ones are the affected ones. We assign them the blues gradient in line below
de_mp_merge$colors <- c(rep("grey", 7), blues)


cl.colors.affected <- c("#08306bff","#3c8cc3ff","grey", "#c6dbefff","grey","grey","grey","grey","grey","#6baed6ff","grey","#07519cff","grey")
                        
p <- DimPlot(object = rna.combined, reduction = "umap_after_harmony_3", pt.size = 0.6, label = F, cols =alpha(cl.colors.affected,1) ,raster=FALSE, label.size = 12) + ggtitle("Most affected clusters from mitopark 15/18 dataset")

plot_filename <- paste0(plot_directory, "/umap_de_mp1518_norm", ".pdf")
  #png(plot_filename, width = 600, height = 600) 
  pdf(plot_filename, width = 9, height = 6)  # Adjust width and height as desired
  print(p)
  dev.off()  
  
  
#venn diagragm for common up and dowmregulated genes
library(VennDiagram)
mycol<- c("#963634ff","#31869bff")
x = list(de_up= de_60hda_up$Gene, de_down= de_60hda_down$Gene)
if (!require(devtools)) install.packages("devtools")
devtools::install_github("yanlinlin82/ggvenn")
library(ggvenn)
up <- ggvenn(
  x, 
  fill_color = mycol,auto_scale = TRUE,
  stroke_size = 0.5, set_name_size = 4
  )
# Add title after initializing the plot
up <- up + ggtitle("Affected genes mouse striatum")
plot_filename <- paste0(plot_directory, "/venn_diag_60ohda", ".pdf")
  #png(plot_filename, width = 600, height = 600) 
  pdf(plot_filename, width = 4, height = 4)  # Adjust width and height as desired
  print(up)
  dev.off()
  
y=list(de_up= de_mp11_up$Gene, de_down= de_mp11_down$Gene)
down<- ggvenn(
  y, 
  fill_color = mycol,auto_scale = TRUE,
  stroke_size = 0.5, set_name_size = 4
  )
# Add title after initializing the plot
down <- down + ggtitle("downregulated genes")
plot_filename <- paste0(plot_directory, "/venn_diag_mp11", ".pdf")
  #png(plot_filename, width = 600, height = 600) 
  pdf(plot_filename, width = 13, height = 6)  # Adjust width and height as desired
  print(down)
  dev.off()  

#volcano plots

clusters <- unique(Idents(rna.combined))

  keyvals <- ifelse(
    de_mp11$avg_log2FC < -0.2, '#31869bff',
      ifelse(de_mp11$avg_log2FC > 0.2, '#963634ff',
        'black'))
  keyvals[is.na(keyvals)] <- 'grey'
  names(keyvals)[keyvals == '#963634ff'] <- 'Upregulated'
  names(keyvals)[keyvals == '#31869bff'] <- 'Dowregulated'
  
  p <- EnhancedVolcano(de_mp11,
                       lab = NA, FCcutoff = 0.5,
                       pCutoff = 10e-6,
                       x = "avg_log2FC", y = "p_val_adj",pointSize = 2.0,
                        labSize = 4.0, colCustom = keyvals) +
    ylim(0,150)+
    xlim(-2,2)+
    labs(x= "avg log2 FC", y="p val adjusted",title = paste("Volcano plots 6OHDA:"))+
    theme(legend.position="none")+
    theme_classic()
   plot_filename <- paste0(plot_directory, "/volcano_mp11_all",  ".pdf")
  pdf(plot_filename, width = 6, height = 4)  # Adjust width and height as desired
  print(p)
  dev.off() 

    keyvals <- ifelse(
    de_mp$avg_log2FC < -0.2, '#31869bff',
      ifelse(de_mp$avg_log2FC > 0.2, '#963634ff',
        'black'))
  keyvals[is.na(keyvals)] <- 'grey'
  names(keyvals)[keyvals == '#963634ff'] <- 'Upregulated'
  names(keyvals)[keyvals == '#31869bff'] <- 'Dowregulated'
  
  p <- EnhancedVolcano(de_mp,
                       lab = NA, FCcutoff = 0.5,
                       pCutoff = 10e-6,
                       x = "avg_log2FC", y = "p_val_adj",pointSize = 2.0,
                        labSize = 4.0, colCustom = keyvals) +
    ylim(0,150)+
    xlim(-2,2)+
    labs(x= "avg log2 FC", y="p val adjusted",title = paste("Volcano plots Mitopark 15- 18 weeks:"))+
    theme(legend.position="none")+
    theme_classic()
   plot_filename <- paste0(plot_directory, "/volcano_mp_all",  ".pdf")
  pdf(plot_filename, width = 6, height = 4)  # Adjust width and height as desired
  print(p)
  dev.off() 

      keyvals <- ifelse(
    de_6ohda$avg_log2FC < -0.2, '#31869bff',
      ifelse(de_6ohda$avg_log2FC > 0.2, '#963634ff',
        'black'))
  keyvals[is.na(keyvals)] <- 'grey'
  names(keyvals)[keyvals == '#963634ff'] <- 'Upregulated'
  names(keyvals)[keyvals == '#31869bff'] <- 'Dowregulated'
  
  p <- EnhancedVolcano(de_6ohda,
                       lab = NA, FCcutoff = 0.5,
                       pCutoff = 10e-6,
                       x = "avg_log2FC", y = "p_val_adj",pointSize = 2.0,
                        labSize = 4.0, colCustom = keyvals) +
    ylim(0,150)+
    xlim(-2,2)+
    labs(x= "avg log2 FC", y="p val adjusted",title = paste("Volcano plots mild 60hda:"))+
    theme(legend.position="none")+
    theme_classic()
   plot_filename <- paste0(plot_directory, "/volcano_6ohda_all",  ".pdf")
  pdf(plot_filename, width = 6, height = 4)  # Adjust width and height as desired
  print(p)
  dev.off() 
#volcano plot per cluster 
  
for (clust in clusters) {
   subset_data_6ohda <- subset(de_6ohda, cluster== clust)
    subset_data_6ohda <- subset_data_6ohda[subset_data_6ohda$p_val_adj < 0.001 ,]
  subset_data_mp <- subset(de_mp, cluster== clust)
  subset_data_mp<-subset_data_mp[subset_data_mp$p_val_adj < 0.0001 , ]
#  GO_terms<- subset(GO_results_common, cluster== clust)
# # Split the geneID column by "/" and unnest the resulting list column
# GO_terms <- GO_terms %>%
#   mutate(geneIDs = strsplit(geneID, "/")) %>%
#   unnest(geneIDs)
# 
# # Keep only unique gene names
# GO_terms <- GO_terms %>%
#   distinct(geneIDs)
  # Replace forward slash with underscore in the cluster name
  clust_filename <- gsub("/", "_", clust)
  keyvals <- ifelse(
    subset_data_6ohda$avg_log2FC < -0.2, '#31869bff',
      ifelse(subset_data_6ohda$avg_log2FC > 0.2, '#963634ff',
        'black'))
  keyvals[is.na(keyvals)] <- 'grey'
  names(keyvals)[keyvals == '#963634ff'] <- 'Upregulated'
  names(keyvals)[keyvals == '#31869bff'] <- 'Dowregulated'

  p <- EnhancedVolcano(subset_data_6ohda,
                       lab = NA, FCcutoff = 0.5,
                       pCutoff = 10e-6,
                       x = "avg_log2FC", y = "p_val_adj",pointSize = 2.0,
                        colCustom = keyvals) +
    ylim(0,100)+
    xlim(-2,2)+
    labs(x= "avg log2 FC", y="p val adjusted",title = paste("Volcano plots for cluster IN 6OHDA:", clust))+
    theme(legend.position="none")+
    theme_classic()
   plot_filename <- paste0(plot_directory, "/volcano_6ohda_col_last_", clust_filename, ".pdf")
  pdf(plot_filename, width = 6, height = 4)  # Adjust width and height as desired
  print(p)
  dev.off()
  keyvals <- ifelse(
    subset_data_mp$avg_log2FC < -0.2, '#31869bff',
      ifelse(subset_data_mp$avg_log2FC > 0.2, '#963634ff',
        'black'))
  keyvals[is.na(keyvals)] <- 'grey'
  names(keyvals)[keyvals =='#963634ff'] <- 'Upregulated'
  names(keyvals)[keyvals =='#31869bff'] <- 'Dowregulated'
  
  p2 <- EnhancedVolcano(subset_data_mp,
                       lab = NA, FCcutoff = 0.5,
                       pCutoff = 10e-6,
                       x = "avg_log2FC", y = "p_val_adj",pointSize = 2.0,
                      colCustom = keyvals) +
        ylim(0,100)+
    xlim(-2,2)+
    labs(x= "avg log2 FC", y="p val adjusted",title = paste("Volcano plots for cluster IN MP:", clust))+
    theme(legend.position="none")+
    theme_classic()

  plot_filename <- paste0(plot_directory, "/volcano_mp_col_last_", clust_filename, ".pdf")
  pdf(plot_filename, width = 6, height = 4)  # Adjust width and height as desired
  print(p2)
  dev.off()
}

```
 


```{r}


        
plot_directory <- "Z:/dmclab/Marta/Combined_snRNAseq/final_plots2"
de_6ohda$Experiment <- "6OHDA"

de_mp$Experiment <- "Mitopark"


m1 <- data.frame(matrix(ncol = 5, nrow = 0))

pb <- progress_bar$new(total = nrow(de_mp))
for (i in 1:nrow(de_mp)){
  pb$tick()
  Sys.sleep(1/nrow(de_mp))
  for (j in 1:nrow(de_6ohda)){
    if (de_mp[i, 1] == de_6ohda[j, 1] && de_mp[i, 7] == de_6ohda[j, 7]){
      m1[nrow(m1)+1, 1] <- de_mp[i, 1]
      m1[nrow(m1), 2] <- de_mp[i, 7]
      m1[nrow(m1), 3] <- de_mp[i, 3]
      m1[nrow(m1), 4] <- de_6ohda[j, 3]
      if (m1[nrow(m1), 3]*m1[nrow(m1), 4] < 0){
        m1[nrow(m1), 5] <- "Divergent"}
       else {
        m1[nrow(m1), 5] <- "Convergent"}
      }
    }
  } 
colnames(m1)<- c("cluster","Gene","avg_log2FC_mp","avg_log2FC_6ohda", "agreement" )
write.xlsx(m1, file = "Z:/dmclab/Marta/Combined_snRNAseq/new_allen_based_clustering/common_genes_DEGs_p0001_thresh0.2.xlsx")
m1<- read_xlsx("Z:/dmclab/Marta/Combined_snRNAseq/new_allen_based_clustering/common_genes_DEGs_p0001_thresh0.2.xlsx")

m1_filter<- filter(m1, agreement== "Convergent")
  
dataset_list<- list("mild 6OHDA"=de_6ohda, "Mitopark 15_18"=de_mp, "Shared"=m1_filter)
clusters<- unique(de_6ohda$cluster)

for (i in names(dataset_list)){
  df<- dataset_list[[i]]
 for (clust in clusters){
   if (i=="Shared"){
    df1<- subset(df, cluster== clust) 
# Define the excluded genes
excluded_genes <- c("Malat1", "Xist", "Uty", "Tsix", "Ddx3y", "Ddx3x", 
                    "Tbl1x", "Usp9x", "Jpx", "Ftx", "Tsnax", "Pcdh11x", 
                    "Eif2s3y", "Eif2s3x", "Tmsb4x", "Tmsb4y", "Kdm5d", "Bicc1")

# Create a regex pattern from the excluded genes
pattern <- paste0(excluded_genes, collapse = "|")

# Apply all filtering conditions in one step
m1_forplots <- df1[
  !grepl(pattern, df1$gene) &  # Exclude specific genes
  !grepl("^Gm", df1$gene) &    # Exclude genes starting with "Gm"
  !grepl("Rik$", df1$gene) &   # Exclude genes ending with "Rik"
  !grepl("Ptgds", df1$gene) &  # Exclude "Ptgds"
  !grepl("Sntn", df1$gene) &   # Exclude "Sntn"
  !grepl("AC149090.1", df1$gene),  # Exclude "AC149090.1"
]

# Calculate a combined score
mean<- cbind(m1_forplots$avg_log2FC_6ohda, m1_forplots$avg_log2FC_mp)
m1_forplots$mean <- rowMeans(mean)

# Get the 10 most negative combined scores
most_negative <- m1_forplots[order(m1_forplots$mean), ][1:14, ]

# Get the 10 most positive combined scores
most_positive <- m1_forplots[order(-m1_forplots$mean), ][1:15, ]
# Combine the results
result <- rbind(most_negative, most_positive)
result<- result[,-c(5,6)]
result1<- result[,-4]
result1$experiment <- "Mitopark"
colnames(result1)<- c("cluster","gene","avg_log2FC","experiment")
result2<- result[,-3]
result2$experiment <- "6OHDA"
colnames(result2)<- c("cluster","gene","avg_log2FC","experiment" )
res_long <- rbind(result1, result2)
plot<- ggplot(res_long, aes(reorder(gene, -avg_log2FC),avg_log2FC, fill= avg_log2FC))+
       geom_bar(stat="identity", width=0.7, position= "dodge2")+
        scale_fill_gradient2(
    low = '#31869bff', 
    mid = 'white',
    high = '#963634ff', 
    midpoint = 0,limits = c(-1, 1), oob = scales::squish 
  ) +
  ylim(-2,2)+
  facet_wrap(.~experiment, scales = "fixed", drop=TRUE)+
       theme_classic()+
  theme(
      axis.text.x = element_text( angle=0, vjust = 0.5, hjust = 0, size=10),
      axis.text.y = element_text(size = 12),
      axis.title.y = element_blank()
    )+
  coord_flip()+
 ggtitle(paste("Most significantly changed genes")) 
 clust <- gsub("/", "_", clust)
 plot_filename <- paste0(plot_directory, "/Most_sign_genes_bar_3_",clust, "_",i,".pdf")
  pdf(plot_filename, width = 7, height = 4)   # Adjust width and height as desired
  print(plot)
  dev.off()  

 }else{
   df1<- subset(df, cluster== clust)
excluded_genes <- c("Malat1", "Xist", "Uty", "Tsix", "Ddx3y", "Ddx3x", 
                    "Tbl1x", "Usp9x", "Jpx", "Ftx", "Tsnax", "Pcdh11x", 
                    "Eif2s3y", "Eif2s3x", "Tmsb4x", "Tmsb4y", "Kdm5d", "Bicc1")

# Create a regex pattern from the excluded genes
pattern <- paste0(excluded_genes, collapse = "|")

# Apply all filtering conditions in one step
m1_forplots <- df1[
  !grepl(pattern, df1$Gene) &  # Exclude specific genes
  !grepl("^Gm", df1$Gene) &    # Exclude genes starting with "Gm"
  !grepl("Rik$", df1$Gene) &   # Exclude genes ending with "Rik"
  !grepl("Ptgds", df1$Gene) &  # Exclude "Ptgds"
  !grepl("Sntn", df1$Gene) &   # Exclude "Sntn"
  !grepl("AC149090.1", df1$Gene),  # Exclude "AC149090.1"
]

most_negative <- m1_forplots[order(unique(m1_forplots$avg_log2FC)), ][1:15, ]
most_positive <- m1_forplots[order(unique(-m1_forplots$avg_log2FC)), ][1:15, ]

# Combine the results
result <- rbind(most_negative, most_positive)
result<- na.omit(result)

# # Calculate the number of elements in each cluster
#   cluster_counts <- result %>%
#     count(cluster) %>%
#     arrange(desc(n))
#   
#   # Reorder clusters based on the number of elements
#   result <- result %>%
#     mutate(cluster = factor(cluster, levels = cluster_counts$cluster))

plot<-ggplot(result, aes(reorder(Gene, -avg_log2FC),avg_log2FC, fill= avg_log2FC))+
       geom_bar(stat="identity", width=0.7)+
        scale_fill_gradient2(
    low = '#31869bff', 
    mid = 'white',
    high = '#963634ff', 
    midpoint = 0,limits = c(-1, 1), oob = scales::squish 
  ) +
  ylim(-2,2)+
  #facet_wrap(.~cluster, scales = "fixed", drop=TRUE)+
       theme_classic()+
  theme(
      axis.text.x = element_text( angle=0, vjust = 0.5, hjust = 0, size=10),
      axis.text.y = element_text(size = 12),
      axis.title.y = element_blank()
    )+
  coord_flip()+
 ggtitle(paste("Most significantly changed genes")) 
clust <- gsub("/", "_", clust)
 plot_filename <- paste0(plot_directory, "/Most_sign_genes_bar_3_",clust, "_",i,".pdf")
  pdf(plot_filename, width = 5, height = 4)   # Adjust width and height as desired
  print(plot)
  dev.off()  
  }
 }
}


dataset_list<- list("Mitopark 11"=de_mp11)
for (i in names(dataset_list)){
  df1<- dataset_list[[i]]

   if (i=="Shared"){
# Define the excluded genes
excluded_genes <- c("Malat1", "Xist", "Uty", "Tsix", "Ddx3y", "Ddx3x", 
                    "Tbl1x", "Usp9x", "Jpx", "Ftx", "Tsnax", "Pcdh11x", 
                    "Eif2s3y", "Eif2s3x", "Tmsb4x", "Tmsb4y", "Kdm5d","Bicc1")

# Create a regex pattern from the excluded genes
pattern <- paste0(excluded_genes, collapse = "|")

# Apply all filtering conditions in one step
m1_forplots <- df1[
  !grepl(pattern, df1$gene) &  # Exclude specific genes
  !grepl("^Gm", df1$gene) &    # Exclude genes starting with "Gm"
  !grepl("Rik$", df1$gene) &   # Exclude genes ending with "Rik"
  !grepl("Ptgds", df1$gene) &  # Exclude "Ptgds"
  !grepl("Sntn", df1$gene) &   # Exclude "Sntn"
  !grepl("AC149090.1", df1$gene),  # Exclude "AC149090.1"
]


# Calculate a combined score
mean<- cbind(m1_forplots$avg_log2FC_6ohda, m1_forplots$avg_log2FC_mp)
m1_forplots$mean <- rowMeans(mean)

# Get the 10 most negative combined scores
most_negative <- m1_forplots[order(m1_forplots$mean), ][1:20, ]

# Get the 10 most positive combined scores
most_positive <- m1_forplots[order(-m1_forplots$mean), ][1:30, ]
# Combine the results
result <- rbind(most_negative, most_positive)
result<- result[,-c(5,6)]
result1<- result[,-4]
result1$experiment <- "Mitopark"
colnames(result1)<- c("cluster","gene","avg_log2FC","experiment")
result2<- result[,-3]
result2$experiment <- "6OHDA"
colnames(result2)<- c("cluster","gene","avg_log2FC","experiment" )
res_long <- rbind(result1, result2)

  plot<- ggplot(res_long, aes( reorder(gene, -avg_log2FC),cluster,fill= avg_log2FC))+
       geom_tile()+
        scale_fill_gradient2(
    low = '#31869bff', 
    mid = 'white',
    high = '#963634ff', 
    midpoint = 0,limits = c(-2, 2), oob = scales::squish 
  ) +
       facet_grid( .~ experiment) +
       theme_classic()+
  theme(
      axis.text.x = element_text( angle=90, vjust = 0.5, hjust = 0, size=10),
      axis.text.y = element_text(size = 12),
      axis.title.y = element_blank()
    ) 
  plot_filename <- paste0(plot_directory, "/Most_sign_div_genes_",i, ".pdf")
  pdf(plot_filename, width = 14, height = 4)  # Adjust width and height as desired
  print(plot)
  dev.off()  
 }else{
 # Define the excluded genes
excluded_genes <- c("Malat1", "Xist", "Uty", "Tsix", "Ddx3y", "Ddx3x", 
                    "Tbl1x", "Usp9x", "Jpx", "Ftx", "Tsnax", "Pcdh11x", 
                    "Eif2s3y", "Eif2s3x", "Tmsb4x", "Tmsb4y", "Kdm5d","Bicc1")

# Create a regex pattern from the excluded genes
pattern <- paste0(excluded_genes, collapse = "|")

# Apply all filtering conditions in one step
m1_forplots <- df1[
  !grepl(pattern, df1$Gene) &  # Exclude specific genes
  !grepl("^Gm", df1$Gene) &    # Exclude genes starting with "Gm"
  !grepl("Rik$", df1$Gene) &   # Exclude genes ending with "Rik"
  !grepl("Ptgds", df1$Gene) &  # Exclude "Ptgds"
  !grepl("Sntn", df1$Gene) &   # Exclude "Sntn"
  !grepl("AC149090.1", df1$Gene),  # Exclude "AC149090.1"
]

   
most_negative<- m1_forplots[order(unique(m1_forplots$avg_log2FC)), ][1:20, ]
most_positive <- m1_forplots[order(unique(-m1_forplots$avg_log2FC)), ][1:30, ]

# Combine the results
result <- rbind(most_negative, most_positive)
result<- na.omit(result)

# # Calculate the number of elements in each cluster
#   cluster_counts <- result %>%
#     count(cluster) %>%
#     arrange(desc(n))
#   
#   # Reorder clusters based on the number of elements
#   result <- result %>%
#     mutate(cluster = factor(cluster, levels = cluster_counts$cluster))


  plot<- ggplot(result, aes( reorder(Gene, -avg_log2FC),cluster,fill= avg_log2FC))+
       geom_tile()+
        scale_fill_gradient2(
    low = '#31869bff', 
    mid = 'white',
    high = '#963634ff', 
    midpoint = 0,limits = c(-2, 2), oob = scales::squish 
  ) +
       theme_classic()+
  theme(
      axis.text.x = element_text( angle=90, vjust = 0.5, hjust = 0, size=10),
      axis.text.y = element_text(size = 12),
      axis.title.y = element_blank()
    ) 
  plot_filename <- paste0(plot_directory, "/Most_sign_div_genes_",i, ".pdf")
  pdf(plot_filename, width = 8, height = 4)  # Adjust width and height as desired
  print(plot)
  dev.off()  
  }
 }


m1_forplots<- subset(m1_forplots, cluster %in% c("Striatum D1", "Striatum D2", "oligodendrocytes"))  
write.xlsx(m1_forplots, file="Z:/dmclab/Marta/Combined_snRNAseq/plot yuvarani.xlsx")
```

```{r}



#let'sonly get significat positive genes
clusters<- unique(de_6ohda$cluster)


GO_common<- data.frame()

for (clust in clusters) {
  subset_data_6ohda <- de_genes_6ohda[[clust]]
  subset_data_6ohda_down <- subset_data_6ohda[subset_data_6ohda$p_val_adj < 0.0001 & subset_data_6ohda$avg_log2FC < 0, ]
  subset_data_6ohda_up <- subset_data_6ohda[subset_data_6ohda$p_val_adj < 0.0001 &  subset_data_6ohda$avg_log2FC > 0, ]
#gene_to_test <- rownames(subset_data[subset_data$p_val_adj < 0.01 & (subset_data$avg_log2FC < -0.3 | subset_data$avg_log2FC > 0.3), ])
 gene_to_test_down <- subset_data_6ohda_down$Gene
GO_results<- enrichGO(gene= gene_to_test_down, OrgDb = "org.Mm.eg.db", keyType = "SYMBOL", ont = "MF") 

GO_results_Df<-as.data.frame(GO_results)
GO_results_Df <- GO_results_Df %>%
  mutate(Bgratio_value = as.numeric(strsplit(BgRatio, "/") %>% sapply(function(x) x[1])))
filtered_GO_results_down <- GO_results_Df %>%
    slice_max(n = 40, order_by = as.numeric(Count))
 filtered_GO_results_down <- filtered_GO_results_down %>% add_column(cluster=clust, reg= "downreg")
 GO_common<- rbind(GO_common, filtered_GO_results_down)
 
 gene_to_test_up <- subset_data_6ohda_up$Gene
GO_results<- enrichGO(gene= gene_to_test_up, OrgDb = "org.Mm.eg.db", keyType = "SYMBOL", ont = "MF") 
GO_results_Df<-as.data.frame(GO_results)
GO_results_Df <- GO_results_Df %>%
  mutate(Bgratio_value = as.numeric(strsplit(BgRatio, "/") %>% sapply(function(x) x[1])))
filtered_GO_results_up <- GO_results_Df %>%
    slice_max(n = 40, order_by = as.numeric(Count))
 filtered_GO_results_up <- filtered_GO_results_up %>% add_column(cluster=clust, reg= "upreg")
 GO_common<- rbind(GO_common, filtered_GO_results_up) 

}
GO_common$Experiment<- "6OHDA"

GO_commonMp<- data.frame()

for (clust in clusters) {
  subset_data_mp <- de_genes_mp[[clust]]
  subset_data_mp_down <- subset_data_mp[subset_data_mp$p_val_adj < 0.0001 & subset_data_mp$avg_log2FC < 0, ]
  subset_data_mp_up <- subset_data_mp[subset_data_mp$p_val_adj < 0.0001 &  subset_data_mp$avg_log2FC > 0, ]
#gene_to_test <- rownames(subset_data[subset_data$p_val_adj < 0.01 & (subset_data$avg_log2FC < -0.3 | subset_data$avg_log2FC > 0.3), ])
 gene_to_test_down <- subset_data_mp_down$Gene
GO_results<- enrichGO(gene= gene_to_test_down, OrgDb = "org.Mm.eg.db", keyType = "SYMBOL", ont = "MF") 

GO_results_Df<-as.data.frame(GO_results)
GO_results_Df <- GO_results_Df %>%
  mutate(Bgratio_value = as.numeric(strsplit(BgRatio, "/") %>% sapply(function(x) x[1])))
filtered_GO_results_down <- GO_results_Df %>%
    slice_max(n = 40, order_by = as.numeric(Count))
 filtered_GO_results_down <- filtered_GO_results_down %>% add_column(cluster=clust, reg= "downreg")
 GO_commonMp<- rbind(GO_commonMp, filtered_GO_results_down)
 
 gene_to_test_up <- subset_data_mp_up$Gene
GO_results<- enrichGO(gene= gene_to_test_up, OrgDb = "org.Mm.eg.db", keyType = "SYMBOL", ont = "MF") 
GO_results_Df<-as.data.frame(GO_results)
GO_results_Df <- GO_results_Df %>%
  mutate(Bgratio_value = as.numeric(strsplit(BgRatio, "/") %>% sapply(function(x) x[1])))
filtered_GO_results_up <- GO_results_Df %>%
    slice_max(n = 40, order_by = as.numeric(Count))
 filtered_GO_results_up <- filtered_GO_results_up %>% add_column(cluster=clust, reg= "upreg")
 GO_commonMp<- rbind(GO_commonMp, filtered_GO_results_up) 

}
GO_commonMp$Experiment<- "Mitopark15_18"
GO_common_genes<- data.frame()

m1_filter<-  subset(m1, agreement == "Convergent")

for (clust in cluster) {
  subset_data_common<-subset(m1_filter, cluster== clust)
  avg<- cbind(subset_data_common$avg_log2FC_6ohda, subset_data_common$avg_log2FC_mp)
subset_data_common$mean <- rowMeans(avg)
  subset_data_mp_down <- subset_data_common[ subset_data_common$mean < 0, ]
  subset_data_mp_up <- subset_data_common[   subset_data_common$mean > 0, ]
#gene_to_test <- rownames(subset_data[subset_data$p_val_adj < 0.01 & (subset_data$avg_log2FC < -0.3 | subset_data$avg_log2FC > 0.3), ])
 gene_to_test_down <- subset_data_mp_down$gene

GO_results<- enrichGO(gene= gene_to_test_down, OrgDb = "org.Mm.eg.db", keyType = "SYMBOL", ont = "MF") 

GO_results_Df<-as.data.frame(GO_results)

if (nrow(GO_results_Df) < 1){
  next  
}else{
  GO_results_Df <- GO_results_Df %>%
  mutate(Bgratio_value = as.numeric(strsplit(BgRatio, "/") %>% sapply(function(x) x[1])))
filtered_GO_results_down <- GO_results_Df %>%
    slice_max(n = 40, order_by = as.numeric(Count))
 filtered_GO_results_down <- filtered_GO_results_down %>% add_column(cluster=clust, reg= "downreg")
 GO_common_genes<- rbind(GO_common_genes, filtered_GO_results_down)
 
 gene_to_test_up <- subset_data_mp_up$gene
GO_results<- enrichGO(gene= gene_to_test_up, OrgDb = "org.Mm.eg.db", keyType = "SYMBOL", ont = "MF") 
GO_results_Df<-as.data.frame(GO_results)
GO_results_Df <- GO_results_Df %>%
  mutate(Bgratio_value = as.numeric(strsplit(BgRatio, "/") %>% sapply(function(x) x[1])))
filtered_GO_results_up <- GO_results_Df %>%
    slice_max(n = 40, order_by = as.numeric(Count))
 filtered_GO_results_up <- filtered_GO_results_up %>% add_column(cluster=clust, reg= "upreg")
 GO_common_genes<- rbind(GO_common_genes, filtered_GO_results_up) 

 }

}
GO_common_genes$Experiment<- "Common"

Gene_ontology<- rbind(GO_common, GO_commonMp)
write.xlsx(Gene_ontology, file = "Z:/dmclab/Marta/Combined_snRNAseq/GO_mostfreq_DEGs_withcommonP0001_pval0.5.xlsx")
Gene_ontology_all<- read.xlsx("Z:/dmclab/Marta/Combined_snRNAseq/GO_mostfreq_DEGs.xlsx")

# Extract common values from the 'ID' column
common_ids <- intersect(GO_commonMp$Description, GO_common$Description)

# Filter df1 and df2 to keep only the rows with common IDs
df1_common <- GO_commonMp[GO_commonMp$Description %in% common_ids, ]
df2_common <- GO_common[GO_common$Description %in% common_ids, ]

Gene_ontology_common<- rbind(df1_common, df2_common)
write.xlsx(Gene_ontology_common, file = "Z:/dmclab/Marta/Combined_snRNAseq/GO_mostfreq_common_across_models_DEGs.xlsx")

# Extract unique values from the 'Description' column in Gene_ontology
all_descriptions <- Gene_ontology$Description

# Extract unique values from the 'Description' column in Gene_ontology_common
common_descriptions <- Gene_ontology_common$Description

# Find descriptions in Gene_ontology that are not present in Gene_ontology_common
in_common <- innerjoin(all_descriptions, common_descriptions)

# Filter Gene_ontology to keep only the rows with descriptions not in Gene_ontology_common
not_in_common_rows <- Gene_ontology[Gene_ontology$Description %in% not_in_common, ]
write.xlsx(not_in_common_rows, file = "Z:/dmclab/Marta/Combined_snRNAseq/GO_mostfreq_not_common_across_models_DEGs.xlsx")
Gene_ontology_common<- read.csv("Z:/dmclab/Marta/Combined_snRNAseq/GO_mostfreq_common_across_models_DEGs_sel.csv")
Gene_ontology<- read_xlsx("Z:/dmclab/Marta/Combined_snRNAseq/GO_mostfreq_DEGs_withcommon_sel.xlsx")

clusters<- c("Cortex", "Striatum D1", "oligodendrocytes", "Striatum D2")
plot_directory <- "Z:/dmclab/Marta/Combined_snRNAseq/final_plots2"
dir.create(plot_directory, showWarnings = FALSE)
Gene_ontology<- na.omit(Gene_ontology)
experiments<- unique(Gene_ontology$Experiment)
library(ggplot2)
library(pheatmap)
install.packages("maditr")
library(maditr)
library(tidyr)
library(EnhancedVolcano)
library(scales)
for (clust in cluster) { 

#description<- subset_common$Description 
#subset_common<-  subset_common %>%
    #slice_max(n = 10, order_by = as.numeric(Count))
subset<- subset(Gene_ontology , cluster== clust )
subset_tot<-  subset %>%
  group_by(Experiment)%>%
    slice_max(n = 10, order_by = as.numeric(Count))


# Determine the global range of 'Count' values across all experiments
global_count_range <- range(subset$Count)

# Function to extract the legend
get_legend <- function(plot) {
  legend <- cowplot::get_legend(plot)
  plot <- plot + theme(legend.position = "right")
  return(list(plot = plot, legend = legend))
}

# Initialize list to store plots
plots <- list()
legend <- NULL

for (i in experiments) {
  subset_exp <- subset(subset, Experiment == i)
  subset_exp <- subset_exp %>%
    slice_max(n = 10, order_by = as.numeric(Count))
  
  plot <- ggplot(subset_exp, aes(x = reorder(Description, -log10(p.adjust)), y = -log10(p.adjust), size = Count, color = reg == "downreg")) +
    geom_point() +
    scale_size_continuous(limits = global_count_range,range=c(3,6)) +
    scale_y_continuous(limits=c(0,7.5))+# Ensure consistent size range
    scale_colour_manual(name = 'Reg', values = setNames(c("#31869bff", "#963634ff"), c(T, F))) +
    labs(title = paste("Gene ontology on cluster:", clust, i), size = "Number of Genes") +
    theme_minimal() +
    theme(
      axis.text.x = element_text(angle = 45, vjust = 0.5, hjust = 1),
      axis.text.y = element_text(size = 12),
      axis.title.y = element_blank()
    ) +
    coord_flip()+
    theme_classic()
  
  # Extract legend from the first plot
  if (is.null(legend)) {
    extracted <- get_legend(plot)
    plot <- extracted$plot
    legend <- extracted$legend
    plot <- plot + theme(legend.position = "right")
  } else {
    plot <- plot + theme(legend.position = "right")
  }
  clust <- gsub("/", "_", clust)

  plot_filename <- paste0(plot_directory, "/GO_dotplot_", clust, "_", i, ".pdf")
  ggsave(plot_filename, plot = plot, width = 10, height = 6)
  
  
  

}
  
}

 




```

Cellchat for LR analysis and cel to cell interaction
```{r}
BiocManager::install("BiocNeighbors")
devtools::install_github("jinworks/CellChat")

devtools::install_github("sqjin/CellChat")
install.packages('NMF')
devtools::install_github("jokergoo/circlize")
devtools::install_github("jokergoo/ComplexHeatmap")
library(CellChat) 
library(NMF)
library(circlize)
library(ComplexHeatmap)

DefaultAssay(rna.combined) <- "RNA" # Ensure the RNA assay is selected

degs <- m1$gene
rna.combined_degs <- subset(rna.combined, features = degs)

data.input <- rna.combined[["RNA"]]@data # normalized data matrix
# For Seurat version >= “5.0.0”, get the normalized data via `seurat_object[["RNA"]]$data`
labels <- Idents(rna.combined)
meta <- rna.combined@meta.data
# Create a data frame for the labels
labels_df <- data.frame(labels = labels, row.names = names(labels))
# Merge the existing metadata with the labels based on row names (cell names)
meta<- merge(meta, labels_df, by = "row.names", all.x = TRUE)

# Set the row names back to the original names after merging
row.names(meta_updated) <- meta_updated$Row.names
meta_updated$Row.names <- NULL  # Remove the Row.names column if not needed

# Check the updated metadata
head(meta_updated)

# Create CellChat object from the Seurat object
cellChat <- createCellChat(object = rna.combined, group.by = "ident", assay = "RNA")
#or
cellchat1 <- createCellChat(object = rna.combined, meta = meta, group.by = "labels") # or replace 'ident' with the cell type annotation you have
cellchat1@meta$samples<- cellchat1@meta$Sample
 # set CellChatDB <- CellChatDB.human if working on the human dataset
interaction_input <- CellChatDB$interaction
complex_input <- CellChatDB$complex
cofactor_input <- CellChatDB$cofactor
geneInfo <- CellChatDB$geneInfo
write.csv(interaction_input, file = "interaction_input_CellChatDB.csv")
write.csv(complex_input, file = "complex_input_CellChatDB.csv")
write.csv(cofactor_input, file = "cofactor_input_CellChatDB.csv")
write.csv(geneInfo, file = "geneInfo_input_CellChatDB.csv")

interaction_input <- read.csv(file = 'interaction_input_CellChatDB.csv', row.names = 1)
complex_input <- read.csv(file = 'complex_input_CellChatDB.csv', row.names = 1)
cofactor_input <- read.csv(file = 'cofactor_input_CellChatDB.csv', row.names = 1)
geneInfo <- read.csv(file = ' geneInfo_input_CellChatDB.csv', row.names = 1)
CellChatDB <- list()
CellChatDB$interaction <- interaction_input
CellChatDB$complex <- complex_input
CellChatDB$cofactor <- cofactor_input
CellChatDB$geneInfo <- geneInfo

CellChatDB.mouse <- CellChatDB
usethis::use_data(CellChatDB.mouse, overwrite = TRUE)

# use a subset of CellChatDB for cell-cell communication analysis
CellChatDB.use <- subsetDB(CellChatDB, search = "Secreted Signaling", key = "annotation") # use Secreted Signaling

# Only uses the Secreted Signaling from CellChatDB v1
#  CellChatDB.use <- subsetDB(CellChatDB, search = list(c("Secreted Signaling"), c("CellChatDB v1")), key = c("annotation", "version"))

# use all CellChatDB except for "Non-protein Signaling" for cell-cell communication analysis
# CellChatDB.use <- subsetDB(CellChatDB)


# use all CellChatDB for cell-cell communication analysis
# CellChatDB.use <- CellChatDB # simply use the default CellChatDB. We do not suggest to use it in this way because CellChatDB v2 includes "Non-protein Signaling" (i.e., metabolic and synaptic signaling). 


cellchat <- subsetData(cellchat1) # This step is necessary even if using the whole database
future::plan("multisession", workers = 4) # do parallel
cellchat <- identifyOverExpressedGenes(cellchat)
cellchat <- identifyOverExpressedInteractions(cellchat)
#Compute the communication probability and infer cellular communication network
cellchat <- computeCommunProb(cellchat, type = "triMean")
#Users can filter out the cell-cell communication if there are only few cells in certain cell groups. By default, the minimum number of cells required in each cell group for cell-cell communication is 10.
cellchat <- filterCommunication(cellchat, min.cells = 10)

cellchat <- computeCommunProbPathway(cellchat)
```
