---
title: "figures 6OHDA ST data"
author: "Yuvarani Masarapu"
date: '2024-04-16'
output: html_document
---

```{r}
#X11Fonts()$sans
#[1] "-*-helvetica-%s-%s-*-*-%d-*-*-*-*-*-*-*"
```

# Libraries
```{r lib-load, message=FALSE, warning=FALSE, echo=FALSE}
library(Seurat)
library(SeuratObject)
library(STutility)
library(ggplot2)
library(patchwork)
library(dplyr)
library(data.table)
library(Signac)
library(cowplot)
library(grid)
library(GenomeInfoDb)
library(dplyr)
library(Biobase)
library(BiocManager)
library(AnnotationDbi)
```

# Objects
```{r eval=FALSE}
spatial.data <- readRDS("~/Documents/parkinsonss_figures_folder/sub-clustering/data.integrated_subset_intfeat_3.rds")

indir <- getwd()

colors.cl <- c("#F18F01", "#878787", "#FCCDE5", "#048BA8", "#2E4057", "#99C24D", "#B95F89", "#5F070C" ,  "#DFC27D", "#DB6C79", "#B191FF", "#157A6E", "#73683B", "#97C8EB","#C51B7D", "#BA9763", "#31081F", "#52D1DC", "#700353", "#F5D3C8", "#725752", "#D8315B", "#6B4E71", "#8CBFCF" , "#C2B2B4", "#EDE5A6", "#52B788", "#EE1B26", "#F2F230", "#91F291", "#386CB0", "#E7298A")
```

# UMAP clusters

## Whole dataset
```{r}
p <- DimPlot(object = spatial.data, reduction = "umap_after_harmony", group.by = "seurat_clusters", label = F, pt.size = 0.5, cols = colors.cl)
ggsave(plot = p, filename = paste(indir, "/final_figures/clusters/sub-clusters_ST_6OHDA_UMAP.pdf", sep = ""), height = 10, width = 12, dpi = 300)
```

## Split by condition
```{r}
p <- DimPlot(object = spatial.data, reduction = "umap_after_harmony", group.by = "seurat_clusters", label = T, pt.size = 0.5, split.by = "sample_type", cols = colors.cl)
ggsave(plot = p, filename = paste(indir, "/final_figures/clusters/sub-clusters_ST_6OHDA_UMAP_splitby_condition.pdf", sep = ""), height = 10, width = 20, dpi = 300)
```

# Violins for clusters
```{r}
p <- VlnPlot(object = spatial.data, features = "nCount_Spatial", group.by = "seurat_clusters", pt.size = 0, assay = "Spatial", cols = colors.cl) + ggtitle("UMI/spot for 6OHDA ST subclusters") + NoLegend()

p1 <- VlnPlot(object = spatial.data, features = "nFeature_Spatial", group.by = "seurat_clusters", pt.size = 0, assay = "Spatial", cols = colors.cl) + ggtitle("Genes/spot for 6OHDA ST subclusters")

p2 <- p - p1

p2

ggsave(p2, filename = paste(indir, "/final_figures/clusters/stats_perSpot_unnormalised_groupby_sub-clusters.pdf", sep = ""), dpi = 300, width = 20, height = 6)
```

# Spatial heatmaps of clusters

## with tissue background

```{r fig.height=16, fig.width=16}
images <- c("slice1", "slice1.1.1", "slice1.2.2", "slice1.3.3", "slice1.4.4", "slice1.5.5", "slice1.6.6", "slice1.7.7", "slice1.8.8", "slice1.9.9", "slice1.10.10", "slice1.11.11",
            "slice1.12.12", "slice1.13.13", "slice1.14.14", "slice1.15.15", "slice1.16.16", "slice1.17.17", "slice1.18.18", "slice1.19.19", "slice1.20.20", "slice1.21.21", "slice1.22.22", "slice1.23.23")
samples <- unique(spatial.data$sample_id)
names(colors.cl[1:14]) <- levels(spatial.data)

for(i in 1:length(images)){
  p <- SpatialPlot(object = spatial.data, pt.size.factor = 1, crop = FALSE, image.alpha = 0.8, label = FALSE, group.by = "seurat_clusters", images = images[i], cols = colors.cl) + ggtitle(samples[i])
ggsave(plot = p, filename = paste(indir, "/final_figures/clusters/clusters_on_tissue/spatial_plots_sub-clusters_on_tissue", samples[i], ".pdf", sep = ""), dpi = 300, width = 8, height = 8)
}
```

## without tissue background

```{r fig.height=16, fig.width=16}
images <- c("slice1", "slice1.1.1", "slice1.2.2", "slice1.3.3", "slice1.4.4", "slice1.5.5", "slice1.6.6", "slice1.7.7", "slice1.8.8", "slice1.9.9", "slice1.10.10", "slice1.11.11",
            "slice1.12.12", "slice1.13.13", "slice1.14.14", "slice1.15.15", "slice1.16.16", "slice1.17.17", "slice1.18.18", "slice1.19.19", "slice1.20.20", "slice1.21.21", "slice1.22.22", "slice1.23.23")
samples <- unique(spatial.data$sample_id)
names(colors.cl) <- levels(spatial.data)

for(i in 1:length(images)){
  p <- SpatialPlot(object = spatial.data, pt.size.factor = 1, crop = FALSE, image.alpha = 0, label = TRUE, group.by = "seurat_clusters", images = images[i], cols = colors.cl) + ggtitle(samples[i])
ggsave(plot = p, filename = paste(indir, "/final_figures/clusters/clusters_without_tissue/spatial_plots_sub-clusters_no-tissue_", samples[i], ".pdf", sep = ""), dpi = 300, width = 8, height = 8)
}
```


# Subset ST data

## DE genes from sn data for mapping on ST data

```{r}
library(readxl)
genes_from_marta <- read_excel("~/Documents/deconvolution_OHDA/figures/plot_yuvarani.xlsx")
```

```{r}
heatmap.cl <- c("#E0E0E0", "#BABABA", "#FDDBC7", "#F4A582", "#D6604D", "#B2182B","#67001F")
```

```{r}
dir.create(path = paste(indir, "/sn-genes_for_sub-clusters_with-tissue", sep = ""), showWarnings = TRUE, recursive = FALSE, mode = "0777")
```


```{r}
genes <- genes_from_marta$gene
genes <- union(genes, c("Nr4a1", "Chd7", "Etv5", "Slc24a3", "Hdac4", "Mafg", "Neto2", "Zdhhc14", "Unc13a", "Grip1", "Tgfbr3", "Smg5", "Tmod3", "Slc24a4", "Sv2c", "Spata13", "Per3", "Sun2", "Pcsk1n", "Apod"))
```

### with tissue background

```{r fig.height=16, fig.width=16}
images <- c("slice1", "slice1.1.1", "slice1.2.2", "slice1.3.3", "slice1.4.4", "slice1.5.5", "slice1.6.6", "slice1.7.7", "slice1.8.8", "slice1.9.9", "slice1.10.10", "slice1.11.11",
            "slice1.12.12", "slice1.13.13", "slice1.14.14", "slice1.15.15", "slice1.16.16", "slice1.17.17", "slice1.18.18", "slice1.19.19", "slice1.20.20", "slice1.21.21", "slice1.22.22", "slice1.23.23")
samples <- unique(spatial.data$sample_id)

for(i in 1:length(genes)){
  dir.create(path = paste(indir, "/sn-genes_for_sub-clusters_2024-07-26/", genes[i], "/spatial-heatmaps", sep = ""), showWarnings = TRUE, recursive = TRUE, mode = "0777")
  plots <- list()
  for(j in 1:length(images)){
    plots[[j]] <- SpatialFeaturePlot(object = spatial.data, features = genes[i], pt.size.factor = 1.2, stroke = 0, crop = TRUE, image.alpha = 0.5, images = images[j]) +
    ggtitle(samples[j]) +  
    scale_fill_gradientn(colors = heatmap.cl) +
    theme(legend.key.size = unit(2, units = "mm"),
          legend.text = element_text(size = 3),
          legend.title = element_text(size = 5),
          plot.title = element_text(size = 3))
  }
  p <- cowplot::plot_grid(plotlist = plots, nrow = 5, ncol = 5)
  ggsave(plot = p, filename = paste(indir, "/sn-genes_for_sub-clusters_2024-07-26/", genes[i], "/spatial-heatmaps/", genes[i], "_spatial-heatmaps.pdf", sep = ""), dpi = 300, width = 210, height = 297, units = "mm")
}
```

#### UMAP
```{r}
for(i in 1:length(genes)){
  dir.create(path = paste(indir, "/sn-genes_for_sub-clusters_with-tissue/", genes[i], "/umap", sep = ""), showWarnings = TRUE, recursive = TRUE, mode = "0777")
  p <- FeaturePlot(object = spatial.data, reduction = "umap_after_harmony", features = genes[i], pt.size = 0.2, cols = heatmap.cl)
  ggsave(plot = p, filename = paste(indir, "/sn-genes_for_sub-clusters_with-tissue/", genes[i], "/umap/", genes[i], ".pdf", sep = ""), height = 10, width = 12, dpi = 300)
}
```

### without tissue background

```{r}
for(i in 1:length(genes)){
  dir.create(path = paste(indir, "/sn-genes_for_sub-clusters/", genes[i], "/spatial-heatmaps", sep = ""), showWarnings = TRUE, recursive = TRUE, mode = "0777")
  plots <- list()
  for(j in 1:length(images)){
    plots[[j]] <- SpatialFeaturePlot(object = spatial.data, features = genes[i], pt.size.factor = 1.2, stroke = 0, crop = FALSE, image.alpha = 0, images = images[j]) +
    ggtitle(samples[j]) +  
    scale_fill_gradientn(colors = heatmap.cl) +
    theme(legend.key.size = unit(2, units = "mm"),
          legend.text = element_text(size = 3),
          legend.title = element_text(size = 5),
          plot.title = element_text(size = 3))
  }
  p <- cowplot::plot_grid(plotlist = plots, nrow = 5, ncol = 5)
  ggsave(plot = p, filename = paste(indir, "/sn-genes_for_sub-clusters/", genes[i], "/spatial-heatmaps/", genes[i], "_spatial-heatmaps.pdf", sep = ""), dpi = 300, width = 210, height = 297, units = "mm")
}
```

#### UMAP

```{r}
for(i in 1:length(genes)){
  dir.create(path = paste(indir, "/sn-genes_for_sub-clusters/", genes[i], "/umap", sep = ""), showWarnings = TRUE, recursive = TRUE, mode = "0777")
  p <- FeaturePlot(object = spatial.data, reduction = "umap_after_harmony", features = genes[i], pt.size = 0.2, cols = heatmap.cl)
  ggsave(plot = p, filename = paste(indir, "/sn-genes_for_sub-clusters/", genes[i], "/umap/", genes[i], ".pdf", sep = ""), height = 10, width = 12, dpi = 300)
}
```

#### split UMAP by condition

```{r}
genes <- union(genes_from_marta$gene, c("Nr4a1", "Chd7", "Etv5", "Slc24a3", "Hdac4", "Mafg", "Neto2", "Zdhhc14", "Unc13a", "Grip1", "Tgfbr3", "Smg5", "Tmod3", "Slc24a4", "Sv2c", "Spata13", "Per3", "Sun2", "Pcsk1n", "Apod"))
genes <- union(genes, c("Nr4a1", "Dusp14", "Egr1", "Egr4", "Arc", "Nr4a3"))
genes <- unique(genes)

heatmap.cl <- c("#E0E0E0", "#BABABA", "#FDDBC7", "#F4A582", "#D6604D", "#B2182B","#67001F")

indir <- getwd()
for(i in 1:length(genes)){
  p <- FeaturePlot(object = spatial.data, reduction = "umap_after_harmony", features = genes[i], pt.size = 0.2, cols = heatmap.cl, split.by = "sample_type") & theme(legend.position = c(0.9,0.9))
  ggsave(plot = p, filename = paste(indir, "/umap_split_DE_genes/", genes[i], ".pdf", sep = ""), height = 10, width = 20, dpi = 300)
}
```

# UMAP for Nr4a1 in subclusters - spots colored by the order of logFC for Nr4a1 (-ve to +ve direction)

Grey colored spots mean that the gene either didn't show up in that cluster or is not significant (adjusted p-value greater than 0.1)

Nr4a1 
LogFC subclusterID adjusted_p_value
-0,276665745 Subcluster0 2,23037E-14
-0,293148396 Subcluster1 5,21991E-27
-0,528172649 Subcluster2 1,04787E-72
NA Subcluster3 NA
NA Subcluster4 NA
NA Subcluster5 NA
-0,405979549 Subcluster6 3,54394E-12
-0,266244548 Subcluster7 0,027116089
-0,353638494 Subcluster8 0,000893187
-0,33118422 Subcluster9 0,063862427
-0,673243771 Subcluster10 8,35905E-12
-0,447822803 Subcluster11 0,087419319
NA Subcluster12 NA
NA Subcluster13 NA

```{r}
logfc_df_Nr4a1 <- data.frame(logFC = c(-0.276665745, -0.293148396, -0.528172649, NA, NA, NA, -0.405979549, -0.266244548, -0.353638494, -0.33118422, -0.673243771, -0.447822803, NA, NA),
                             subclusters = levels(spatial.data))


logfc_df_Nr4a1 <- logfc_df_Nr4a1[order(logfc_df_Nr4a1$logFC, decreasing = TRUE), ] #because the values are all -ve, Nr4a1 is down

#Check for colors that can help visualise the direction of expression
RColorBrewer::display.brewer.all()
#RColorBrewer::display.brewer.pal(n=9, name = "Blues")
reds <- RColorBrewer::brewer.pal(n = 9, name = "Reds") #we will take all the 9 colors from here
#blues <- RColorBrewer::brewer.pal(n = 9, name = "Blues") #we will take all the 9 colors from here

#logFC_colors_Nr4a1 <- c(blues , rep("grey", 5)) #Now here we have our 14 colors, first 9 ordered according to logFC direction as lightest to darkest, and rest 5 subclusters are grey meaning no DE expression.
logFC_colors_Nr4a1 <- c(reds , rep("grey", 5))

logfc_df_Nr4a1$logFC_colors <- logFC_colors_Nr4a1 #add this to the dataframe

logfc_df_Nr4a1$subclusters <- as.numeric(logfc_df_Nr4a1$subclusters) #Now we reorder the subclusters so we have 0 to 13.
#The colors will also be reordered and then we use the new colors order for the UMAP. We do this because in UMAP the colors mapped to clusters as per the cluster order of 0 to 13.
logfc_df_Nr4a1 <- logfc_df_Nr4a1[order(logfc_df_Nr4a1$subclusters, decreasing = FALSE), ]
```

```{r}
#for UMAP we need the colors mapping from above
logfc_df_Nr4a1$logFC_colors
```

```{r}
indir <- getwd()
p <- DimPlot(object = spatial.data, group.by = "seurat_clusters", reduction = "umap_after_harmony", pt.size = 0.5, label = F, cols = logfc_df_Nr4a1$logFC_colors, label.size = 12) + ggtitle("logFC direction for DE gene Nr4a1 in subclusters")
ggsave(plot = p, filename = paste(indir, "/Nr4a1_logFC_sections_BLUES/final_figures/UMAP_Nr4a1-logFC_in_subclusters.pdf", sep = ""), height = 10, width = 12, dpi = 300)
```

## Plot the same on tissue sections in background

```{r fig.height=16, fig.width=16}
images <- c("slice1", "slice1.1.1", "slice1.2.2", "slice1.3.3", "slice1.4.4", "slice1.5.5", "slice1.6.6", "slice1.7.7", "slice1.8.8", "slice1.9.9", "slice1.10.10", "slice1.11.11",
            "slice1.12.12", "slice1.13.13", "slice1.14.14", "slice1.15.15", "slice1.16.16", "slice1.17.17", "slice1.18.18", "slice1.19.19", "slice1.20.20", "slice1.21.21", "slice1.22.22", "slice1.23.23")
samples <- unique(spatial.data$sample_id)
colors.subcl <- logfc_df_Nr4a1$logFC_colors
names(colors.subcl) <- levels(spatial.data)

#for(i in 1:length(images)){
#  p <- SpatialPlot(object = spatial.data, pt.size.factor = 1, crop = FALSE, image.alpha = 0.8, label = TRUE, group.by = "seurat_clusters", images #= images[i], cols = colors.subcl) + ggtitle(samples[i])
#ggsave(plot = p, filename = paste(getwd(), "/Nr4a1_logFC_sections_BLUES/Nr4a1_logFC-change_on-", samples[i], ".pdf", sep = ""), dpi = 300, width = 8, height = 8)
#}

for(i in 1:length(images)){
  p <- SpatialPlot(object = spatial.data, pt.size.factor = 1, crop = FALSE, image.alpha = 0.8, label = TRUE, group.by = "seurat_clusters", images = images[i], cols = colors.subcl) + ggtitle(samples[i])
ggsave(plot = p, filename = paste(getwd(), "/Nr4a1_logFC_sections_REDS/Nr4a1_logFC-change_on-", samples[i], ".pdf", sep = ""), dpi = 300, width = 8, height = 8)
}
```

# UMAP for Nr4a1 in MAIN CLUSTERs - spots colored by the order of logFC for Nr4a1 (-ve to +ve direction)

Grey colored spots mean that the gene either didn't show up in that cluster or is not significant (adjusted p-value greater than 0.1)

Nr4a1 
LogFC clusterID adjusted_p_value
-0,276665745 Subcluster0 2,23037E-14
-0,293148396 Subcluster1 5,21991E-27
-0,528172649 Subcluster2 1,04787E-72
NA Subcluster3 NA
NA Subcluster4 NA
NA Subcluster5 NA
-0,405979549 Subcluster6 3,54394E-12
-0,266244548 Subcluster7 0,027116089
-0,353638494 Subcluster8 0,000893187
-0,33118422 Subcluster9 0,063862427
-0,673243771 Subcluster10 8,35905E-12
-0,447822803 Subcluster11 0,087419319
NA Subcluster12 NA
NA Subcluster13 NA

```{r}
library(readxl)
logfc_df_Nr4a1 <- data.frame()
count <- 0
for(i in 0:31){
  genes.de <- read_excel("/Users/yuvarani.masarapu/Documents/parkinsonss_figures_folder/DE_genes_control_vs_treated_per_cluster.xlsx", sheet = as.character(i))
  if("Nr4a1" %in% genes.de$gene){
    logfc.val <- genes.de$avg_log2FC[genes.de$gene %in% "Nr4a1"]
    count <- count + 1
  }
  else{
    logfc.val <- NA
  }
  tmp <- data.frame(logFC = logfc.val, clusters = i)
  logfc_df_Nr4a1 <- rbind(logfc_df_Nr4a1, tmp)
  
}

logfc_df_Nr4a1 <- logfc_df_Nr4a1[order(logfc_df_Nr4a1$logFC, decreasing = TRUE), ] #because the values are all -ve, Nr4a1 is down

#clusters between -0.2 to -0.4
#clusters between -0.4 to -0.6
#clusters above -0.6

#Check for colors that can help visualise the direction of expression
RColorBrewer::display.brewer.all()
reds <- RColorBrewer::brewer.pal(n = 3, name = "Reds") #we will take all the 9 colors from here

colors.fc <- c()
for(i in 1:length(logfc_df_Nr4a1$logFC)){
  if(is.na(logfc_df_Nr4a1$logFC[i])){
    colors.fc[i] <- "grey"
  }
  else 
    {
      if((logfc_df_Nr4a1$logFC[i] > -0.3) | (logfc_df_Nr4a1$logFC[i] == -0.3)){
        colors.fc[i] <- reds[1]
      }
      else if((logfc_df_Nr4a1$logFC[i] < -0.3) & (logfc_df_Nr4a1$logFC[i] > -0.5)){
        colors.fc[i] <- reds[2]
      }
      else if((logfc_df_Nr4a1$logFC[i] < -0.5) | (logfc_df_Nr4a1$logFC[i] == -0.5)){
        colors.fc[i] <- reds[3]
      }
    }
}
```

```{r}
spatial.data <- readRDS("~/Documents/parkinsonss_figures_folder/data.integrated.clusters_2022-08-22.rds")

p <- DimPlot(object = spatial.data, group.by = "seurat_clusters", reduction = "umap_after_harmony", pt.size = 0.5, label = F, cols = colors.fc, label.size = 12) + ggtitle("logFC direction for DE gene Nr4a1 in main clusters")
ggsave(plot = p, filename = "/Users/yuvarani.masarapu/Documents/parkinsonss_figures_folder/UMAP_Nr4a1-logFC_in_main_clusters.pdf", height = 10, width = 12, dpi = 300)
```

## Plot the same on tissue sections in background

```{r fig.height=16, fig.width=16}
images <- c("slice1", "slice1.2", "slice1.2", "slice1.3", "slice1.4", "slice1.5", "slice1.6", "slice1.7", "slice1.8", "slice1.9", "slice1.10", "slice1.11",
            "slice1.12", "slice1.13", "slice1.14", "slice1.15", "slice1.16", "slice1.17", "slice1.18", "slice1.19", "slice1.20", "slice1.21", "slice1.22", "slice1.23")
samples <- unique(spatial.data$sample_id)
names(colors.fc) <- levels(spatial.data)

for(i in 1:length(images)){
  p <- SpatialPlot(object = spatial.data, pt.size.factor = 1, crop = FALSE, image.alpha = 0.8, label = TRUE, group.by = "seurat_clusters", images = images[i], cols = colors.fc) + ggtitle(samples[i])
ggsave(plot = p, filename = paste("/Users/yuvarani.masarapu/Documents/parkinsonss_figures_folder/Nr4a1_logFC-change_on-", samples[i], "main_clusters.pdf", sep = ""), dpi = 300, width = 8, height = 8)
}
```


# UMAP for Egr1 in subclusters - spots colored by the order of logFC for Nr4a1 (-ve to +ve direction)

Egr1 
LogFC subclusterID adjusted_p_value
-0,26059349 Subcluster0 1,20758E-09
-0,25934584 Subcluster1 8,82449E-13
-0,39590644 Subcluster2 1,68218E-35
NA Subcluster3 NA
NA Subcluster4 NA
NA Subcluster5 NA
-0,427916651 Subcluster6 8,92119E-14
NA Subcluster7 NA
NA Subcluster8 NA
-0,432890587 Subcluster9 0,000344567
-0,417768413 Subcluster10 0,000695712
NA Subcluster11 NA
NA Subcluster12 NA
NA Subcluster13 NA

```{r}
logfc_df <- data.frame(logFC = c(-0.26059349, -0.25934584, -0.39590644, NA, NA, NA, -0.427916651, NA, NA, -0.432890587, -0.417768413, NA, NA, NA),
                       subclusters = levels(spatial.data))


logfc_df <- logfc_df[order(logfc_df$logFC, decreasing = TRUE), ] #because the values are all -ve, Egr1 is down

#Here we need 6 colors that increase in intensity to mimic the logFC direction change of Egr1 and rest are all grey clusters

#Check for colors that can help visualise the direction of expression
#RColorBrewer::display.brewer.pal(n=6, name = "Reds")
reds <- RColorBrewer::brewer.pal(n = 6, name = "Reds") #we will take all the 9 colors from here
#blues <- RColorBrewer::brewer.pal(n = 6, name = "Blues")

#logFC_colors <- c(blues , rep("grey", 8)) #Now we have the 14 colors we need
logFC_colors <- c(reds , rep("grey", 8))

logfc_df$logFC_colors <- logFC_colors #add this to the dataframe

logfc_df$subclusters <- as.numeric(logfc_df$subclusters) #Now we reorder the subclusters so we have 0 to 13.
#The colors will also be reordered and then we use the new colors order for the UMAP. We do this because in UMAP the colors mapped to clusters as per the cluster order of 0 to 13.
logfc_df <- logfc_df[order(logfc_df$subclusters, decreasing = FALSE), ]
```

```{r}
#for UMAP we need the colors mapping from above
logfc_df$logFC_colors
```

```{r}
indir <- getwd()
p <- DimPlot(object = spatial.data, group.by = "seurat_clusters", reduction = "umap_after_harmony", pt.size = 0.5, label = F, cols = logfc_df$logFC_colors, label.size = 12) + ggtitle("logFC direction for DE gene Egr1 in subclusters")
ggsave(plot = p, filename = paste(indir, "/final_figures/UMAP_Egr1-logFC_in_subclusters.pdf", sep = ""), height = 10, width = 12, dpi = 300)
```

## Plot on tissue background 

```{r fig.height=16, fig.width=16}
images <- c("slice1", "slice1.1.1", "slice1.2.2", "slice1.3.3", "slice1.4.4", "slice1.5.5", "slice1.6.6", "slice1.7.7", "slice1.8.8", "slice1.9.9", "slice1.10.10", "slice1.11.11",
            "slice1.12.12", "slice1.13.13", "slice1.14.14", "slice1.15.15", "slice1.16.16", "slice1.17.17", "slice1.18.18", "slice1.19.19", "slice1.20.20", "slice1.21.21", "slice1.22.22", "slice1.23.23")
samples <- unique(spatial.data$sample_id)
colors.subcl <- logfc_df$logFC_colors
names(colors.subcl) <- levels(spatial.data)

#for(i in 1:length(images)){
#  p <- SpatialPlot(object = spatial.data, pt.size.factor = 1, crop = FALSE, image.alpha = 0.8, label = TRUE, group.by = "seurat_clusters", images #= images[i], cols = colors.subcl) + ggtitle(samples[i])
#ggsave(plot = p, filename = paste(getwd(), "/Egr1_logFC_sections_BLUES/Egr1_logFC-change_on-", samples[i], ".pdf", sep = ""), dpi = 300, width = 8, height = 8)
#}

for(i in 1:length(images)){
  p <- SpatialPlot(object = spatial.data, pt.size.factor = 1, crop = FALSE, image.alpha = 0.8, label = TRUE, group.by = "seurat_clusters", images = images[i], cols = colors.subcl) + ggtitle(samples[i])
ggsave(plot = p, filename = paste(getwd(), "/Egr1_logFC_sections_REDS/Egr1_logFC-change_on-", samples[i], ".pdf", sep = ""), dpi = 300, width = 8, height = 8)
}
```

# UMAP for Dusp14 and Arc showing the logFC change across the affected clusters

Dusp14

LogFC subclusterID adjusted_p_value
NA Subcluster0 NA
-0,25934584 Subcluster1 8,82449E-13
-0,39590644 Subcluster2 1,68218E-35
NA Subcluster3 NA
NA Subcluster4 NA
NA Subcluster5 NA
-0,427916651 Subcluster6 8,92119E-14
NA Subcluster7 NA
NA Subcluster8 NA
-0,432890587 Subcluster9 0,000344567
-0,417768413 Subcluster10 0,000695712
NA Subcluster11 NA
NA Subcluster12 NA
NA Subcluster13 NA

# Panel 4G - gene expression plots with expression divided into labels instead of continous values

```{r}
spatial.data <- readRDS("~/Documents/parkinsonss_figures_folder/sub-clustering/data.integrated_subset_intfeat_3.rds")
```

### Test labels for one gene as example

```{r}
subset.Nr4a1 <- subset(spatial.data, features = "Nr4a1")
#subset.Nr4a1 <- subset.Nr4a1[ , colSums(subset.Nr4a1) > 0]
exp.df <- data.frame(value = subset.Nr4a1@assays$SCT@data[1:15242])
low = floor(min(subset.Nr4a1@assays$SCT@data[1:15242]))
high = ceiling(max(subset.Nr4a1@assays$SCT@data[1:15242]))
#5 groups of colors
#let's say low, medium, medium high, high
#low, low+(high/4), (high/2), high-(high/4), high
#0.75,1.5,2.25,3

exp.df$labels <- ifelse((exp.df$value == low), "No-exp", "NA")
exp.df$labels[(exp.df$value > low) & (exp.df$value < (low+(high/4)))] <- "low"
exp.df$labels[((exp.df$value == (low+(high/4)) | exp.df$value > (low+(high/4))) & (exp.df$value < (high/2)))] <- "medium"
exp.df$labels[((exp.df$value == (high/2) | exp.df$value > (high/2)) & (exp.df$value < (high-(high/4))))] <- "medium high"
exp.df$labels[((exp.df$value == (high-(high/4)) | exp.df$value > (high-(high/4))) & (exp.df$value < high))] <- "high"
subset.Nr4a1$exp_labels <- exp.df$labels
#subset.Nr4a1$exp_labels[subset.Nr4a1@assays$SCT@data[]]


images <- c("slice1", "slice1.1.1", "slice1.2.2", "slice1.3.3", "slice1.4.4", "slice1.5.5", "slice1.6.6", "slice1.7.7", "slice1.8.8", "slice1.9.9", "slice1.10.10", "slice1.11.11",
            "slice1.12.12", "slice1.13.13", "slice1.14.14", "slice1.15.15", "slice1.16.16", "slice1.17.17", "slice1.18.18", "slice1.19.19", "slice1.20.20", "slice1.21.21", "slice1.22.22", "slice1.23.23")

#colors.labels <- c("lightgrey", "#FDDBC7", "#F4A582", "#B2182B", "#67001F")
colors.labels <- c(greys[2] ,reds[2], reds[4], reds[7], reds[9])
names(colors.labels) <- c("No-exp","low", "medium", "medium high", "high")
SpatialPlot(object = subset.Nr4a1, pt.size.factor = 1.2, crop = FALSE, image.alpha = 1, label = FALSE, group.by = "exp_labels", images = c(images[1], images[5]), cols = colors.labels)

reds <- RColorBrewer::brewer.pal(9, "Reds")
RColorBrewer::display.brewer.pal(9, "Reds")
RColorBrewer::display.brewer.all()
greys <- RColorBrewer::brewer.pal(2, "Greys")
```

## Loop with colorscale and keeping spots with no-expression

```{r}
reds <- RColorBrewer::brewer.pal(9, "Reds")
greys <- RColorBrewer::brewer.pal(2, "Greys")

images <- c("slice1", "slice1.1.1", "slice1.2.2", "slice1.3.3", "slice1.4.4", "slice1.5.5", "slice1.6.6", "slice1.7.7", "slice1.8.8", "slice1.9.9", "slice1.10.10", "slice1.11.11",
            "slice1.12.12", "slice1.13.13", "slice1.14.14", "slice1.15.15", "slice1.16.16", "slice1.17.17", "slice1.18.18", "slice1.19.19", "slice1.20.20", "slice1.21.21", "slice1.22.22", "slice1.23.23")
samples <- unique(spatial.data$sample_id)
library(readxl)
genes_from_marta <- read_excel("~/Documents/deconvolution_OHDA/figures/plot_yuvarani.xlsx")

genes <- unique(genes_from_marta$gene)
#genes from sn-data
genes_sn <- c("Nr4a1", "Chd7", "Etv5", "Slc24a3", "Hdac4", "Mafg", "Neto2", "Zdhhc14", "Unc13a", "Grip1", "Tgfbr3", "Smg5", "Tmod3", "Slc24a4", "Sv2c", "Spata13", "Per3", "Sun2", "Pcsk1n", "Apod", "Egr1")
genes <- union(genes, genes_sn) #combine them both for the final list

colors.labels <- c(greys[2] ,reds[2], reds[4], reds[7], reds[9])
names(colors.labels) <- c("No-exp","low", "medium", "medium high", "high")

indir <- getwd()
for(i in 1:length(genes)){
  dir.create(path = paste(indir, "/gene-expression-plots_method1_grey-to-red-colors/", genes[i], "/spatial-heatmaps", sep = ""), showWarnings = TRUE, recursive = TRUE, mode = "0777")
  
  subset.Nr4a1 <- subset(spatial.data, features = genes[i]) #subset object for the gene
  exp.df <- data.frame(value = subset.Nr4a1@assays$SCT@data[1:length(colnames(subset.Nr4a1))]) #move expression values for the gene into a dataframe for easy manipulation
  low = floor(min(subset.Nr4a1@assays$SCT@data[1:length(colnames(subset.Nr4a1))])) #find the round off of the lowest expression value
  high = ceiling(max(subset.Nr4a1@assays$SCT@data[1:length(colnames(subset.Nr4a1))])) #find the round off of the highest expression value
  #5 groups of colors
  #let's say No exp, low, medium, medium high, high
  #low, low+(high/4), (high/2), high-(high/4), high
  #Example 0.75,1.5,2.25,3
  
  #Make a new column in the dataframe that gives labels according to the expression values
  exp.df$labels <- ifelse((exp.df$value == low), "No-exp", "NA")
  exp.df$labels[(exp.df$value > low) & (exp.df$value < (low+(high/4)))] <- "low"
  exp.df$labels[((exp.df$value == (low+(high/4)) | exp.df$value > (low+(high/4))) & (exp.df$value < (high/2)))] <- "medium"
  exp.df$labels[((exp.df$value == (high/2) | exp.df$value > (high/2)) & (exp.df$value < (high-(high/4))))] <- "medium high"
  exp.df$labels[((exp.df$value == (high-(high/4)) | exp.df$value > (high-(high/4))) & (exp.df$value < high))] <- "high"
  
  #assign these new labels back to the seurat object as metadata
  subset.Nr4a1$exp_labels <- exp.df$labels

  #we now plot spatial plots based on these labels
  plots <- list()
  for(j in 1:length(images)){
    plots[[j]] <- SpatialPlot(object = subset.Nr4a1, pt.size.factor = 1.2, crop = FALSE, image.alpha = 1, label = FALSE, group.by = "exp_labels", images = images[j], cols = colors.labels) +
    ggtitle(samples[j]) +  
    #scale_fill_gradientn(colors = heatmap.cl) +
    theme(legend.key.size = unit(2, units = "mm"),
          legend.text = element_text(size = 3),
          legend.title = element_text(size = 5),
          plot.title = element_text(size = 3))
  }
  p <- cowplot::plot_grid(plotlist = plots, nrow = 5, ncol = 5)
  ggsave(plot = p, filename = paste(indir, "/gene-expression-plots_method1_grey-to-red-colors/", genes[i], "/spatial-heatmaps/", genes[i], "_spatial-heatmaps.pdf", sep = ""), dpi = 300, width = 210, height = 297, units = "mm")
}
```

## Loop with colorscale and removing spots with no-expression

```{r}
images <- c("slice1", "slice1.1.1", "slice1.2.2", "slice1.3.3", "slice1.4.4", "slice1.5.5", "slice1.6.6", "slice1.7.7", "slice1.8.8", "slice1.9.9", "slice1.10.10", "slice1.11.11",
            "slice1.12.12", "slice1.13.13", "slice1.14.14", "slice1.15.15", "slice1.16.16", "slice1.17.17", "slice1.18.18", "slice1.19.19", "slice1.20.20", "slice1.21.21", "slice1.22.22", "slice1.23.23")
library(readxl)
genes_from_marta <- read_excel("~/Documents/deconvolution_OHDA/figures/plot_yuvarani.xlsx")
getwd()

genes <- unique(genes_from_marta$gene)
#genes from sn-data
genes_sn <- c("Nr4a1", "Chd7", "Etv5", "Slc24a3", "Hdac4", "Mafg", "Neto2", "Zdhhc14", "Unc13a", "Grip1", "Tgfbr3", "Smg5", "Tmod3", "Slc24a4", "Sv2c", "Spata13", "Per3", "Sun2", "Pcsk1n", "Apod")
genes <- union(genes, genes_sn) #combine them both for the final list

colors.labels <- c(reds[2], reds[4], reds[7], reds[9]) #colorscale changed
names(colors.labels) <- c("low", "medium", "medium high", "high") #assign colors to named vector so the color to label mapping is correct

for(i in 1:length(genes)){
  dir.create(path = paste(indir, "/gene-expression-plots_method2_grey-to-red-colors/", genes[i], "/spatial-heatmaps", sep = ""), showWarnings = TRUE, recursive = TRUE, mode = "0777")
  
  subset.Nr4a1 <- subset(spatial.data, features = genes[i]) #subset object for the gene
  exp.df <- data.frame(value = subset.Nr4a1@assays$SCT@data[1:length(colnames(subset.Nr4a1))]) #move expression values for the gene into a dataframe for easy manipulation
  low = floor(min(subset.Nr4a1@assays$SCT@data[1:length(colnames(subset.Nr4a1))])) #find the round off of the lowest expression value
  high = ceiling(max(subset.Nr4a1@assays$SCT@data[1:length(colnames(subset.Nr4a1))])) #find the round off of the highest expression value
  #5 groups of colors
  #let's say No exp, low, medium, medium high, high
  #low, low+(high/4), (high/2), high-(high/4), high
  #Example 0.75,1.5,2.25,3
  
  #Make a new column in the dataframe that gives labels according to the expression values
  exp.df$labels <- ifelse((exp.df$value == low), "No-exp", "NA")
  exp.df$labels[(exp.df$value > low) & (exp.df$value < (low+(high/4)))] <- "low"
  exp.df$labels[((exp.df$value == (low+(high/4)) | exp.df$value > (low+(high/4))) & (exp.df$value < (high/2)))] <- "medium"
  exp.df$labels[((exp.df$value == (high/2) | exp.df$value > (high/2)) & (exp.df$value < (high-(high/4))))] <- "medium high"
  exp.df$labels[((exp.df$value == (high-(high/4)) | exp.df$value > (high-(high/4))) & (exp.df$value < high))] <- "high"
  
  #assign these new labels back to the seurat object as metadata
  subset.Nr4a1$exp_labels <- exp.df$labels
  subset.Nr4a1 <- subset(subset.Nr4a1, cells = colnames(subset.Nr4a1)[subset.Nr4a1$exp_labels %in% c("low", "medium", "medium high", "high")]) #remove the spots with no expression

  #we now plot spatial plots based on these labels
  plots <- list()
  for(j in 1:length(images)){
    plots[[j]] <- SpatialPlot(object = subset.Nr4a1, pt.size.factor = 1.2, crop = FALSE, image.alpha = 1, label = FALSE, group.by = "exp_labels", images = images[j], cols = colors.labels) +
    ggtitle(samples[j]) +  
    #scale_fill_gradientn(colors = heatmap.cl) +
    theme(legend.key.size = unit(2, units = "mm"),
          legend.text = element_text(size = 3),
          legend.title = element_text(size = 5),
          plot.title = element_text(size = 3))
  }
  p <- cowplot::plot_grid(plotlist = plots, nrow = 5, ncol = 5)
  ggsave(plot = p, filename = paste(indir, "/gene-expression-plots_method2_grey-to-red-colors/", genes[i], "/spatial-heatmaps/", genes[i], "_spatial-heatmaps.pdf", sep = ""), dpi = 300, width = 210, height = 297, units = "mm")
}
```

# Th expression 

## Loop with colorscale and keeping spots with no-expression

```{r}
images <- c("slice1", "slice1.1.1", "slice1.2.2", "slice1.3.3", "slice1.4.4", "slice1.5.5", "slice1.6.6", "slice1.7.7", "slice1.8.8", "slice1.9.9", "slice1.10.10", "slice1.11.11",
            "slice1.12.12", "slice1.13.13", "slice1.14.14", "slice1.15.15", "slice1.16.16", "slice1.17.17", "slice1.18.18", "slice1.19.19", "slice1.20.20", "slice1.21.21", "slice1.22.22", "slice1.23.23")
indir <- getwd()

genes <- c("Th")

colors.labels <- c(greys[2] ,reds[2], reds[4], reds[7], reds[9])
names(colors.labels) <- c("No-exp","low", "medium", "medium high", "high")

for(i in 1:length(genes)){
  dir.create(path = paste(indir, "/", genes[i] ,"_method1_subclustering", sep = ""), showWarnings = TRUE, recursive = TRUE, mode = "0777")
  
  subset.Nr4a1 <- subset(spatial.data, features = genes[i]) #subset object for the gene
  exp.df <- data.frame(value = subset.Nr4a1@assays$SCT@data[1:length(colnames(subset.Nr4a1))]) #move expression values for the gene into a dataframe for easy manipulation
  low = floor(min(subset.Nr4a1@assays$SCT@data[1:length(colnames(subset.Nr4a1))])) #find the round off of the lowest expression value
  high = ceiling(max(subset.Nr4a1@assays$SCT@data[1:length(colnames(subset.Nr4a1))])) #find the round off of the highest expression value
  #5 groups of colors
  #let's say No exp, low, medium, medium high, high
  #low, low+(high/4), (high/2), high-(high/4), high
  #Example 0.75,1.5,2.25,3
  
  #Make a new column in the dataframe that gives labels according to the expression values
  exp.df$labels <- ifelse((exp.df$value == low), "No-exp", "NA")
  exp.df$labels[(exp.df$value > low) & (exp.df$value < (low+(high/4)))] <- "low"
  exp.df$labels[((exp.df$value == (low+(high/4)) | exp.df$value > (low+(high/4))) & (exp.df$value < (high/2)))] <- "medium"
  exp.df$labels[((exp.df$value == (high/2) | exp.df$value > (high/2)) & (exp.df$value < (high-(high/4))))] <- "medium high"
  exp.df$labels[((exp.df$value == (high-(high/4)) | exp.df$value > (high-(high/4))) & (exp.df$value < high))] <- "high"
  
  #assign these new labels back to the seurat object as metadata
  subset.Nr4a1$exp_labels <- exp.df$labels

  #we now plot spatial plots based on these labels
  plots <- list()
  for(j in 1:length(images)){
    plots[[j]] <- SpatialPlot(object = subset.Nr4a1, pt.size.factor = 1.2, crop = FALSE, image.alpha = 1, label = FALSE, group.by = "exp_labels", images = images[j], cols = colors.labels) +
    ggtitle(samples[j]) +  
    #scale_fill_gradientn(colors = heatmap.cl) +
    theme(legend.key.size = unit(2, units = "mm"),
          legend.text = element_text(size = 3),
          legend.title = element_text(size = 5),
          plot.title = element_text(size = 3))
  }
  p <- cowplot::plot_grid(plotlist = plots, nrow = 5, ncol = 5)
  ggsave(plot = p, filename = paste(indir, "/", genes[i] ,"_method1_subclustering/", genes[i], "_subclustering_method1.pdf", sep = ""), dpi = 300, width = 210, height = 297, units = "mm")
}
```

## Loop with colorscale and removing spots with no-expression

```{r}
images <- c("slice1", "slice1.1.1", "slice1.2.2", "slice1.3.3", "slice1.4.4", "slice1.5.5", "slice1.6.6", "slice1.7.7", "slice1.8.8", "slice1.9.9", "slice1.10.10", "slice1.11.11",
            "slice1.12.12", "slice1.13.13", "slice1.14.14", "slice1.15.15", "slice1.16.16", "slice1.17.17", "slice1.18.18", "slice1.19.19", "slice1.20.20", "slice1.21.21", "slice1.22.22", "slice1.23.23")
indir <- getwd()

genes <- c("Th")

colors.labels <- c(reds[2], reds[4], reds[7], reds[9]) #colorscale changed
names(colors.labels) <- c("low", "medium", "medium high", "high") #assign colors to named vector so the color to label mapping is correct

for(i in 1:length(genes)){
  dir.create(path = paste(indir, "/", genes[i] ,"_method2_subclustering", sep = ""), showWarnings = TRUE, recursive = TRUE, mode = "0777")
  
  subset.Nr4a1 <- subset(spatial.data, features = genes[i]) #subset object for the gene
  exp.df <- data.frame(value = subset.Nr4a1@assays$SCT@data[1:length(colnames(subset.Nr4a1))]) #move expression values for the gene into a dataframe for easy manipulation
  low = floor(min(subset.Nr4a1@assays$SCT@data[1:length(colnames(subset.Nr4a1))])) #find the round off of the lowest expression value
  high = ceiling(max(subset.Nr4a1@assays$SCT@data[1:length(colnames(subset.Nr4a1))])) #find the round off of the highest expression value
  #5 groups of colors
  #let's say No exp, low, medium, medium high, high
  #low, low+(high/4), (high/2), high-(high/4), high
  #Example 0.75,1.5,2.25,3
  
  #Make a new column in the dataframe that gives labels according to the expression values
  exp.df$labels <- ifelse((exp.df$value == low), "No-exp", "NA")
  exp.df$labels[(exp.df$value > low) & (exp.df$value < (low+(high/4)))] <- "low"
  exp.df$labels[((exp.df$value == (low+(high/4)) | exp.df$value > (low+(high/4))) & (exp.df$value < (high/2)))] <- "medium"
  exp.df$labels[((exp.df$value == (high/2) | exp.df$value > (high/2)) & (exp.df$value < (high-(high/4))))] <- "medium high"
  exp.df$labels[((exp.df$value == (high-(high/4)) | exp.df$value > (high-(high/4))) & (exp.df$value < high))] <- "high"
  
  #assign these new labels back to the seurat object as metadata
  subset.Nr4a1$exp_labels <- exp.df$labels
  subset.Nr4a1 <- subset(subset.Nr4a1, cells = colnames(subset.Nr4a1)[subset.Nr4a1$exp_labels %in% c("low", "medium", "medium high", "high")]) #remove the spots with no expression

  #we now plot spatial plots based on these labels
  plots <- list()
  for(j in 1:length(images)){
    plots[[j]] <- SpatialPlot(object = subset.Nr4a1, pt.size.factor = 1.2, crop = FALSE, image.alpha = 1, label = FALSE, group.by = "exp_labels", images = images[j], cols = colors.labels) +
    ggtitle(samples[j]) +  
    #scale_fill_gradientn(colors = heatmap.cl) +
    theme(legend.key.size = unit(2, units = "mm"),
          legend.text = element_text(size = 3),
          legend.title = element_text(size = 5),
          plot.title = element_text(size = 3))
  }
  p <- cowplot::plot_grid(plotlist = plots, nrow = 5, ncol = 5)
  ggsave(plot = p, filename = paste(indir, "/", genes[i] ,"_method2_subclustering/", genes[i], "_subclustering_method2.pdf", sep = ""), dpi = 300, width = 210, height = 297, units = "mm")
}
```




```{r}

```

```{r}
heatmap.cl <- c("#E0E0E0", "#BABABA", "#FDDBC7", "#F4A582", "#D6604D", "#B2182B","#67001F")
```

```{r}
dir.create(path = paste(indir, "/sn-genes_for_sub-clusters_with-tissue", sep = ""), showWarnings = TRUE, recursive = FALSE, mode = "0777")
```


```{r}
genes <- genes_from_marta$gene
genes <- union(genes, c("Nr4a1", "Chd7", "Etv5", "Slc24a3", "Hdac4", "Mafg", "Neto2", "Zdhhc14", "Unc13a", "Grip1", "Tgfbr3", "Smg5", "Tmod3", "Slc24a4", "Sv2c", "Spata13", "Per3", "Sun2", "Pcsk1n", "Apod"))
```

### without tissue background

```{r}
for(i in 1:length(genes)){
  dir.create(path = paste(indir, "/sn-genes_for_sub-clusters/", genes[i], "/spatial-heatmaps", sep = ""), showWarnings = TRUE, recursive = TRUE, mode = "0777")
  plots <- list()
  for(j in 1:length(images)){
    plots[[j]] <- SpatialFeaturePlot(object = spatial.data, features = genes[i], pt.size.factor = 1.2, stroke = 0, crop = FALSE, image.alpha = 0, images = images[j]) +
    ggtitle(samples[j]) +  
    scale_fill_gradientn(colors = heatmap.cl) +
    theme(legend.key.size = unit(2, units = "mm"),
          legend.text = element_text(size = 3),
          legend.title = element_text(size = 5),
          plot.title = element_text(size = 3))
  }
  p <- cowplot::plot_grid(plotlist = plots, nrow = 5, ncol = 5)
  ggsave(plot = p, filename = paste(indir, "/sn-genes_for_sub-clusters/", genes[i], "/spatial-heatmaps/", genes[i], "_spatial-heatmaps.pdf", sep = ""), dpi = 300, width = 210, height = 297, units = "mm")
}
```

# Same as Panel E (significant genes by total genes) for subclusters

We show the most affected clusters and genecount here that way.

```{r}
#variable_features_5000_USED
indir <- getwd()
library(readxl)

de_genes <- read_excel(paste(indir, "/variable_features_5000_USED/DE_genes_control_vs_treated_per_subcluster.xlsx", sep = ""), sheet = "0")
de_genes$cluster <- c(rep("0", length(de_genes$gene)))
for(i in 1:(length(levels(spatial.data))-1)){
  temp <- read_excel(paste(indir, "/variable_features_5000_USED/DE_genes_control_vs_treated_per_subcluster.xlsx", sep = ""), sheet = i)
  temp$cluster <- c(rep(i, length(temp$gene)))
  temp <- temp[temp$p_val_adj < 0.0001, ]
  
  de_genes <- rbind(de_genes, temp)
}

#de_genes <- de_genes[de_genes$avg_log2FC > 0.3 | de_genes$avg_log2FC < -0.3, ]
```

```{r}
length(unique(de_genes$gene))

min(de_genes$avg_log2FC[de_genes$avg_log2FC > 0])
max(de_genes$avg_log2FC[de_genes$avg_log2FC < 0])
```


### Treated group
```{r}
table_gene_count<- data.frame()
clusters <- c("0","1","2","3","4","5","6","7","8","9", "10", "11", "12", "13")
for (clust in clusters){
   subset <- spatial.data[,spatial.data$sample_type %in% "Treated"]
   subset_cluster <- subset[, subset$seurat_clusters %in% clust]
   subset_cluster <- subset_cluster[rowSums(subset_cluster) > 0, ]
   gene_count<- length(rownames(subset_cluster))
   cluster <- clust
   gene_counts_df<- as.data.frame(gene_count, cluster)
   table_gene_count<- rbind(table_gene_count,gene_counts_df)
   
   rm(subset)
   rm(subset_cluster)
 }

de <- de_genes
table_gene_count$sample_type <- "Treated"
cluster <- rownames(table_gene_count)
table_gene_count<- cbind(cluster, table_gene_count)
colnames(table_gene_count)<- c("cluster","total_genes","sample_type")
de_1 <- de %>% group_by(cluster) %>% summarise(count = n())
de_merge <- merge(de_1, table_gene_count, by = "cluster")
de_merge <- transform(de_merge, norm = count / total_genes)
#de_6ohda_merge <- merge(de_6ohda_merge, table_colors, by = "cluster")

de_merge.treated <- de_merge
# Create the bar plot
p <- ggplot(de_merge.treated, aes(x = reorder(cluster, norm), y = norm)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(x = "Striatal clusters", y = "Significant changed genes/ total genes in the sub-cluster (p value<0.0001)", title = "Genes affected in 6OHDA striatal sub-clusters Treated group") +
  #ylim(0,0.0001)+
  scale_fill_discrete(name = "Cluster")+
  theme(axis.text.x = element_text(angle = 45))+
    coord_flip()+
  theme_classic()
plot_filename <- paste0(getwd(), "/barplot_6OHDA_affected-sub-clusters_treated",  ".pdf")
pdf(plot_filename, width = 6, height = 4)  # Adjust width and height as desired
print(p)
dev.off()
```

### Control group
```{r}
table_gene_count<- data.frame()
clusters <- c("0","1","2","3","4","5","6","7","8","9", "10", "11", "12", "13")
 for (clust in clusters){
   subset <- spatial.data[,spatial.data$sample_type %in% "Control"]
   subset_cluster <- subset[, subset$seurat_clusters %in% clust]
   subset_cluster <- subset_cluster[rowSums(subset_cluster) > 0, ]
   gene_count<- length(rownames(subset_cluster))
   cluster <- clust
   gene_counts_df<- as.data.frame(gene_count, cluster)
   table_gene_count<- rbind(table_gene_count,gene_counts_df)
   
   rm(subset)
   rm(subset_cluster)
 }


de <- de_genes
table_gene_count$sample_type <- "Control"
cluster <- rownames(table_gene_count)
table_gene_count<- cbind(cluster, table_gene_count)
colnames(table_gene_count)<- c("cluster","total_genes","sample_type")
de_1 <- de %>% group_by(cluster) %>% summarise(count = n())
de_merge <- merge(de_1, table_gene_count, by = "cluster")
de_merge <- transform(de_merge, norm = count / total_genes)
#de_6ohda_merge <- merge(de_6ohda_merge, table_colors, by = "cluster")

de_merge.control <- de_merge
# Create the bar plot
p <- ggplot(de_merge, aes(x = reorder(cluster, norm), y = norm)) +
  geom_bar(stat = "identity", position = "dodge") +
  labs(x = "Striatal sub-clusters", y = "Significant changed genes/ total genes in the cluster (p value<0.0001)", title = "Genes affected in 6OHDA striatal sub-clusters Control group") +
  ylim(0,0.00075)+
  scale_fill_discrete(name = "Cluster")+
  theme(axis.text.x = element_text(angle = 45))+
    coord_flip()+
  theme_classic()
plot_filename <- paste0(getwd(), "/barplot_6OHDA_affected-sub-clusters_control",  ".pdf")
pdf(plot_filename, width = 6, height = 4)  # Adjust width and height as desired
print(p)
dev.off()
```


# Same as Panel E (significant genes by total genes) for main clusters

We show the most affected clusters and genecount here that way.

```{r}
indir <- getwd()
library(readxl)

library(readxl)
de_genes <- read_excel("/Users/yuvarani.masarapu/Documents/parkinsonss_figures_folder/DE_genes_control_vs_treated_per_cluster.xlsx", 
    sheet = "0")
de_genes$cluster <- c(rep("0", length(de_genes$gene)))
for(i in 1:(length(levels(spatial.data))-1)){
  temp <- read_excel("/Users/yuvarani.masarapu/Documents/parkinsonss_figures_folder/DE_genes_control_vs_treated_per_cluster.xlsx", sheet = i)
  temp$cluster <- c(rep(i, length(temp$gene)))
  temp <- temp[temp$p_val_adj < 0.0001, ]
  
  de_genes <- rbind(de_genes, temp)
}
```

```{r}
length(unique(de_genes$gene))

min(de_genes$avg_log2FC[de_genes$avg_log2FC > 0])
max(de_genes$avg_log2FC[de_genes$avg_log2FC < 0])
```


### Treated group
```{r}
table_gene_count <- data.frame()
clusters <- c("0","1","2","3","4","5","6","7","8","9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "23", "24", "25", "26", "27", "28", "29", "30", "31")
for (clust in clusters){
   subset_cluster <- spatial.data[, spatial.data$seurat_clusters %in% clust]
   subset_cluster <- subset_cluster[rowSums(subset_cluster) > 0, ]
   gene_count<- length(rownames(subset_cluster))
   cluster <- clust
   gene_counts_df<- as.data.frame(gene_count, cluster)
   table_gene_count<- rbind(table_gene_count,gene_counts_df)
   
   rm(subset)
   rm(subset_cluster)
 }

de <- de_genes
cluster <- rownames(table_gene_count)
table_gene_count<- cbind(cluster, table_gene_count)
colnames(table_gene_count)<- c("cluster","total_genes")
de_1 <- de %>% group_by(cluster) %>% summarise(count = n())
de_merge <- merge(de_1, table_gene_count, by = "cluster")
de_merge <- transform(de_merge, norm = count / total_genes)

RColorBrewer::display.brewer.all()
#RColorBrewer::display.brewer.pal(n=9, name = "Blues")
blues <- RColorBrewer::brewer.pal(n = 9, name = "Blues") #we will take all the 9 colors from here

de_merge <- de_merge[order(de_merge$norm, decreasing = FALSE), ]
de_merge$colors <- c(rep("grey", 19), blues)

de_merge$cluster <- as.numeric(de_merge$cluster)
de_merge <- de_merge[order(de_merge$cluster, decreasing = FALSE), ]

cl.colors.affected <- c(de_merge$colors[1:15], "grey", de_merge$colors[16:18], "grey", de_merge$colors[19:28] , "grey", "grey")
```

Now we plot the clusters on UMAP and tissue

Only the 9 top affected clusters are shown
```{r}
p <- DimPlot(object = spatial.data, group.by = "seurat_clusters", reduction = "umap_after_harmony", pt.size = 0.5, label = F, cols = cl.colors.affected, label.size = 12) + ggtitle("Most affected clusters from full 6OHDA dataset")
ggsave(plot = p, filename = "/Users/yuvarani.masarapu/Documents/parkinsonss_figures_folder/UMAP_most_affected_clusters_6OHDA.pdf", height = 10, width = 12, dpi = 300)
```

```{r fig.height=16, fig.width=16}
images <- c("slice1", "slice1.2", "slice1.2", "slice1.3", "slice1.4", "slice1.5", "slice1.6", "slice1.7", "slice1.8", "slice1.9", "slice1.10", "slice1.11",
            "slice1.12", "slice1.13", "slice1.14", "slice1.15", "slice1.16", "slice1.17", "slice1.18", "slice1.19", "slice1.20", "slice1.21", "slice1.22", "slice1.23")
samples <- unique(spatial.data$sample_id)
names(cl.colors.affected) <- levels(spatial.data)

for(i in 1:length(images)){
  p <- SpatialPlot(object = spatial.data, pt.size.factor = 1, crop = FALSE, image.alpha = 0.8, label = TRUE, group.by = "seurat_clusters", images = images[i], cols = cl.colors.affected) + ggtitle(samples[i])
ggsave(plot = p, filename = paste("/Users/yuvarani.masarapu/Documents/parkinsonss_figures_folder/most_affected_clusters/most_affected_clusters_", samples[i], ".pdf", sep = ""), dpi = 300, width = 8, height = 8)
}
```